---
title: "Readings"
format: html
---

## Readings by day and topic

For each topic, there will be a required reading (or pre-class activity) and several optional readings.

We expect you to have read the required reading before the day. There will not be time during the week of the course to read new materials and therefore we *highly suggest* that you do these readings in advance of the course starting.

### Sunday

#### Basic R

##### Required

If you are new to R, please attempt the following tutorials on your own. If you are experienced in R, skim through the tutorials and make sure that all the topics are familiar to you.

- [Ravinet's introduction to R](https://evolutionarygenetics.github.io/Introduction.html)

- [Going further with R](https://evolutionarygenetics.github.io/Chapter2.html)

Through this course, most of the examples you are likely to encounter will predominately use base R. However, learning Tidyverse syntax will make your life easier in the long run, especially if you are going to be manipulating spatial data.


#### Basic Bash

##### Required

##### Optional

### Monday

#### Land, river, sea

##### Required

- Storfer, A., Patton, A., & Fraik, A. K. (2018). Navigating the interface between landscape genetics and landscape genomics. _Front Genet, 9_, 68. doi:10.3389/fgene.2018.00068

##### Optional

- Manel, S., Schwartz, M. K., Luikart, G., & Taberlet, P. (2003). Landscape genetics: combining landscape ecology and population genetics. Trends in Ecology and Evolution, 18(4), 189-197. doi:papers3://publication/doi/10.1016/S0169-5347(03)00008-9 *This is the original paper defining landscape genetics as a field.*

- Find a good terrestrial overview 

- Selkoe, K. A., Scribner, K. T., & Galindo, H. M. (2016). Waterscape genetics – applications of landscape genetics to rivers, lakes, and seas. In N. Balkenhol, S. A. Cushman, A. Storfer, & L. P. Waits (Eds.), Landscape Genetics: Concepts, Methods, Applications (pp. 1-27): John Wiley & Sons, Ltd. *Includes overviews on some of the contrasting features of marine and freshwater environments*

- Liggins, L., Treml, E. A., & Riginos, C. (2019). Seascape genomics: contextualizing adaptive and neutral genomic variation in the ocean environment.
 *Overviews on issues particular to marine studies*
 
- Blanchet, S., Prunier, J. G., Paz-Vinas, I., Saint-Pe, K., Rey, O., Raffard, A., Mathieu-Begne, E., Loot, G., Fourtune, L., & Dubut, V. (2020). A river runs through it: The causes, consequences, and management of intraspecific diversity in river networks. Evolutionary Applications, 13(6), 1195-1213. doi:10.1111/eva.12941 *Features and emerging conclusions from freshwater systems*


#### Maps & spatial data

##### Required

No readings are required. 

##### Optional


### Tuesday

#### Genetic differentiation

##### Required

##### Optional

#### RDA as a general toolkit

##### Required

- Capblancq, T., & Forester, B. R. (2021). Redundancy analysis: A Swiss Army Knife for landscape genomics. Methods in Ecology and Evolution. doi:10.1111/2041-210x.13722 *Clear explanations and worked examples that we will follow in class*

##### Optional

- Rellstab, C., Gugerli, F., Eckert, A. J., Hancock, A. M., & Holderegger, R. (2015). A practical guide to environmental association analysis in landscape genomics. Molecular Ecology, 24(17), 4348-4370. doi:papers3://publication/doi/10.1111/mec.13322 *Highly recommend reading as a general overview to GEA/EEA analyses and concepts*

- Forester, B. R., Jones, M. R., Joost, S., Landguth, E. L., & Lasky, J. R. (2016). Detecting spatial genetic signatures of local adaptation in heterogeneous landscapes. Molecular Ecology, 25(1), 104-120. doi:papers3://publication/doi/10.1111/mec.13476 *Evaluates various methods for finding genotype-by-environment associations and finds that RDA methods often best*

- Lasky, J. R., Josephs, E. B., & Morris, G. P. (2023). Genotype-environment associations to reveal the molecular basis of environmental adaptation. _Plant Cell, 35_, 125-138. doi:10.1093/plcell/koac267 *Focusing on plants, reviews the strengths and shortcomings of GEA approaches*


### Wednesday

#### Gene flow analyses

##### Required

##### Optional

### Thursday

#### Resistance surfaces

##### Required

##### Optional

#### Future projections with GDM & GFs

##### Required

- Fitzpatrick, M. C., & Keller, S. R. (2015). Ecological genomics meets community-level modelling of biodiversity: mapping the genomic landscape of current and future environmental adaptation. Ecol Lett, 18(1), 1-16. doi:papers3://publication/doi/10.1111/ele.12376


##### Optional

- Mokany, K., Ware, C., Woolley, S. N. C., Ferrier, S., Fitzpatrick, Matthew C., & Bahn, V. (2022). A working guide to harnessing generalized dissimilarity modelling for biodiversity analysis and conservation assessment. _Global Ecology and Biogeography, 31_(4), 802-821. doi:10.1111/geb.13459. *A worked example for GDM*


- Ferrier, S., Manion, G., Elith, J., & Richardson, K. (2007). Using generalized dissimilarity modelling to analyse and predict patterns of beta diversity in regional biodiversity assessment. _Diversity and Distributions, 13_(3), 252-264. doi:papers3://publication/doi/10.1111/j.1472-4642.2007.00341.x *Original method for GDM*

- Ellis, N., Smith, S. J., & Pitcher, C. R. (2012). Gradient forests: calculating importance gradients on physical predictors. _Ecology, 93_(1), 156-168. doi:papers3://publication/uuid/06CCD6BA-6A4B-49EB-AE26-736E8A04B9C8 *Original method for GF*


#### Biophysical models

##### Required

- Jahnke, M., Jonsson, P. R., Moksnes, P. O., Loo, L. O., Nilsson Jacobi, M., & Olsen, J. L. (2017). Seascape genetics and biophysical connectivity modelling support conservation of the seagrass *Zostera marina* in the Skagerrak-Kattegat region of the eastern North Sea. Evolutionary Applications, 1-46. doi:papers3://publication/doi/10.1111/eva.12589 *Comprehensive review of studies pairing seascape genetic data with biophysical models*


##### Optional

- Boulanger, E., Dalongeville, A., Andrello, M., Mouillot, D., & Manel, S. (2020). Spatial graphs highlight how multi‐generational dispersal shapes landscape genetic patterns. Ecography. doi:10.1111/ecog.05024


### Friday

#### More GEA: genetic offsets and maladaptation

##### Required

- Capblancq, T., Fitzpatrick, M. C., Bay, R. A., Exposito-Alonso, M., & Keller, S. R. (2020). Genomic prediction of (mal)adaptation across current and future climatic landscapes. _Annual Review of Ecology, Evolution, and Systematics, 51_(1), 245-269. doi:10.1146/annurev-ecolsys-020720-042553 *Review of maladaptation and genomic offset concepts*


##### Optional

- Gougherty, A. V., Keller, S. R., & Fitzpatrick, M. C. (2021). Maladaptation, migration and extirpation fuel climate change risk in a forest tree species. _Nat Clim Change, 11_(2), 166-171. doi:10.1038/s41558-020-00968-6 *Interesting conceptual example that integrates genomic offsets with dispersal capacity*

- Fitzpatrick, M. C., V. E. Chhatre, R. Y. Soolanayakanahally and S. R. Keller, 2021 Experimental support for genomic prediction of climate maladaptation using the machine learning approach Gradient Forests. Mol Ecol Resour 21**:** 2749-2765. *One of the best examples of an empirical test of GEA predictions*


#### Landscape genomics and genetic architectures

##### Required

(None)

##### Optional

- Laruson, A. J., Fitzpatrick, M. C., Keller, S. R., Haller, B. C., & Lotterhos, K. E. (2022). Seeing the forest for the trees: Assessing genetic offset predictions from gradient forest. _Evol Appl, 15_(3), 403-416. doi:10.1111/eva.13354

- Lotterhos, K. E. (2019). The effect of neutral recombination variation on genome scans for selection. _G3 (Bethesda), 9_(6), 1851-1867. doi:10.1534/g3.119.400088

- Bierne, N., Gagnaire, P.-A., & David, P. (2013). The geography of introgression in a patchy environment and the thorn in the side of ecological speciation. _Current Zoology, 59_(1), 72-86. doi:papers3://publication/doi/10.1093/czoolo/59.1.72

