% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
\PassOptionsToPackage{dvipsnames,svgnames,x11names}{xcolor}
%
\documentclass[
  letterpaper,
  DIV=11,
  numbers=noendperiod]{scrartcl}

\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else  
    % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\setlength{\emergencystretch}{3em} % prevent overfull lines
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
% Make \paragraph and \subparagraph free-standing
\ifx\paragraph\undefined\else
  \let\oldparagraph\paragraph
  \renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
  \let\oldsubparagraph\subparagraph
  \renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{241,243,245}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.40,0.45,0.13}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\BuiltInTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\ExtensionTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.28,0.35,0.67}{#1}}
\newcommand{\ImportTok}[1]{\textcolor[rgb]{0.00,0.46,0.62}{#1}}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\NormalTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.07,0.07,0.07}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}

\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother

\KOMAoption{captions}{tableheading}
\makeatletter
\makeatother
\makeatletter
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\AtBeginDocument{%
\ifdefined\contentsname
  \renewcommand*\contentsname{Table of contents}
\else
  \newcommand\contentsname{Table of contents}
\fi
\ifdefined\listfigurename
  \renewcommand*\listfigurename{List of Figures}
\else
  \newcommand\listfigurename{List of Figures}
\fi
\ifdefined\listtablename
  \renewcommand*\listtablename{List of Tables}
\else
  \newcommand\listtablename{List of Tables}
\fi
\ifdefined\figurename
  \renewcommand*\figurename{Figure}
\else
  \newcommand\figurename{Figure}
\fi
\ifdefined\tablename
  \renewcommand*\tablename{Table}
\else
  \newcommand\tablename{Table}
\fi
}
\@ifpackageloaded{float}{}{\usepackage{float}}
\floatstyle{ruled}
\@ifundefined{c@chapter}{\newfloat{codelisting}{h}{lop}}{\newfloat{codelisting}{h}{lop}[chapter]}
\floatname{codelisting}{Listing}
\newcommand*\listoflistings{\listof{codelisting}{List of Listings}}
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\@ifpackageloaded{subcaption}{}{\usepackage{subcaption}}
\makeatother
\makeatletter
\@ifpackageloaded{tcolorbox}{}{\usepackage[skins,breakable]{tcolorbox}}
\makeatother
\makeatletter
\@ifundefined{shadecolor}{\definecolor{shadecolor}{rgb}{.97, .97, .97}}
\makeatother
\makeatletter
\makeatother
\makeatletter
\makeatother
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same} % disable monospaced font for URLs
\hypersetup{
  pdftitle={Slendr simulations},
  pdfauthor={Mark Ravinet},
  colorlinks=true,
  linkcolor={blue},
  filecolor={Maroon},
  citecolor={Blue},
  urlcolor={Blue},
  pdfcreator={LaTeX via pandoc}}

\title{Slendr simulations}
\author{Mark Ravinet}
\date{}

\begin{document}
\maketitle
\ifdefined\Shaded\renewenvironment{Shaded}{\begin{tcolorbox}[frame hidden, sharp corners, enhanced, boxrule=0pt, interior hidden, breakable, borderline west={3pt}{0pt}{shadecolor}]}{\end{tcolorbox}}\fi

\hypertarget{introduction---slendr}{%
\subsection{Introduction - slendr}\label{introduction---slendr}}

\href{https://www.slendr.net/}{\texttt{slendr}} is a recently developed
and extensively well-managed R package that acts as a front end for
demographic simulations with both \texttt{msprime} and \texttt{SLiM}. It
massively simplifies the process of running these programs, making them
easy to interface with via \texttt{R} - so if you're not great at
\texttt{python} (like me!) you can easily use \texttt{R} instead. It is
also very flexible and because it is built on top of the extremely
efficient simulation programs, it is very fast and lightweight.

Much of the information in this tutorial is based on the original
\texttt{slendr}{]}(https://www.slendr.net/) tutorial docs - I would
strongly recommend referring to these as they are the definitive source
for learning how to start to using the package.

\hypertarget{setting-up}{%
\subsection{Setting up}\label{setting-up}}

First we need to load the packages we will need for this session. They
have already been installed, so this will not take long.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(slendr)}
\FunctionTok{library}\NormalTok{(tidyverse)}
\end{Highlighting}
\end{Shaded}

Note that if you install \texttt{slendr} on your own machine, you will
need it to setup all the back-end it uses to interact with python etc.
We will not do this today because it is already configured for you.
However to ensure you know how to do this, you simply run

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{setup\_env}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

Even if \texttt{slendr} is setup, it is always a good idea to initiate
the environment, to ensure everything is working properly ahead of
analyses. Do this like so:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{init\_env}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

Now we're ready to begin simulating!

\hypertarget{a-simple-non-spatial-simulation-with-two-populations}{%
\subsection{A simple non-spatial simulation with two
populations}\label{a-simple-non-spatial-simulation-with-two-populations}}

\texttt{slendr} allows you to combine some simple and logically named R
functions into a complex demographic model. The easiest way to learn how
to do this is to simply use these functions! Here we will simulate two
populations. The second splits from the first 30,000 generations ago.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Population a with 30,000 individuals, arising 50000 generations ago}
\NormalTok{a }\OtherTok{\textless{}{-}} \FunctionTok{population}\NormalTok{(}\StringTok{"a"}\NormalTok{, }\AttributeTok{time =} \DecValTok{50000}\NormalTok{, }\AttributeTok{N =} \DecValTok{30000}\NormalTok{)}

\CommentTok{\# Population b with 15,000 individuals, arising 30,000 generations ago}
\NormalTok{b }\OtherTok{\textless{}{-}} \FunctionTok{population}\NormalTok{(}\StringTok{"b"}\NormalTok{, }\AttributeTok{parent =}\NormalTok{ a, }\AttributeTok{time =} \DecValTok{30000}\NormalTok{, }\AttributeTok{N =} \DecValTok{15000}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Be sure to check what these populations look like in the R environment.
Just call them as objects to see.

Since we are keeping this model very simple in order to learn, all we
need to do next is compile the model - all components are combined
(i.e.~pop size etc) into a single R object.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{model }\OtherTok{\textless{}{-}} \FunctionTok{compile\_model}\NormalTok{(}
  \AttributeTok{populations =} \FunctionTok{list}\NormalTok{(a, b),}
  \AttributeTok{generation\_time =} \DecValTok{1}\NormalTok{,}
  \AttributeTok{direction =} \StringTok{"backward"}\NormalTok{,}
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Note that here we specify the direction of this model is ``backwards'' -
i.e.~we are performing a coalescent (backwards-in-time) simulation.

One really nice feature of \texttt{slendr} is that it allows you to plot
the model to ensure you have specified it correctly.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot\_model}\NormalTok{(model)}
\end{Highlighting}
\end{Shaded}

If everything looks as it should, we are ready to begin simulating.

\hypertarget{running-a-simulation}{%
\subsubsection{Running a simulation}\label{running-a-simulation}}

Next up, we need to set our sampling scheme. Here we will take 30
individuals from each of our two populations:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{samples }\OtherTok{\textless{}{-}} \FunctionTok{schedule\_sampling}\NormalTok{(model, }\AttributeTok{times =} \DecValTok{0}\NormalTok{, }\FunctionTok{list}\NormalTok{(a, }\DecValTok{30}\NormalTok{), }\FunctionTok{list}\NormalTok{(b, }\DecValTok{30}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

Note that this function is called \texttt{schedule\_sampling} - this is
because it allows you to set the timing of sampling (i.e.~you can sample
at different temporal points across the model) and also the location -
actually spatially if you wish. But more on that later.

With this setup, we can run the simulation. We will use \texttt{msprime}
because this is a coalescent simulation over many thousands of
generations and therefore it is much more efficient and fast.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ts }\OtherTok{\textless{}{-}} \FunctionTok{msprime}\NormalTok{(model, }\AttributeTok{samples =}\NormalTok{ samples, }\AttributeTok{sequence\_length =} \DecValTok{1000}\NormalTok{, }\AttributeTok{recombination\_rate =} \DecValTok{0}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

This should only take a few seconds. Note we have simulated a 1000 bp
seuqence here with no recombination, but we could easily alter this if
we wanted to. Next we should take a look at the simulation output.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ts }
\end{Highlighting}
\end{Shaded}

So far this doesn't show a great deal, other than the fact that it is a
tree-sequence stored temporarily on our computer. To get at what the
simulation actually shows, we need to process it.

\hypertarget{processing-simulations---a-simple-example}{%
\subsubsection{Processing simulations - a simple
example}\label{processing-simulations---a-simple-example}}

Right now our simulation output is a tree sequence of 60 individuals, 30
from population a and 30 from population b. But we might want to
calculate some population genetic statistics on this dataset. So here
work towards calculating \emph{F}\textsubscript{ST}.

First of all, we need to set up our populations - i.e.~let the pipeline
know which individuals are in each population. This is easy enough to do
with built in functions and some \texttt{tidyR}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{a\_pop }\OtherTok{\textless{}{-}} \FunctionTok{ts\_samples}\NormalTok{(ts) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{filter}\NormalTok{(pop }\SpecialCharTok{==}\NormalTok{ a) }\SpecialCharTok{\%\textgreater{}\%}\NormalTok{ .}\SpecialCharTok{$}\NormalTok{name}
\NormalTok{b\_pop }\OtherTok{\textless{}{-}} \FunctionTok{ts\_samples}\NormalTok{(ts) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{filter}\NormalTok{(pop }\SpecialCharTok{==}\NormalTok{ b) }\SpecialCharTok{\%\textgreater{}\%}\NormalTok{ .}\SpecialCharTok{$}\NormalTok{name}
\end{Highlighting}
\end{Shaded}

Next we need to add mutations to our tree sequence. One of the reasons
msprime is so fast is that it records only the tree sequence or
genealogy - mutations can be added after the fact in order to ensure the
simulation is extremely efficient. Here we will add simulations using a
basic SNP mutation rate.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ts\_m }\OtherTok{\textless{}{-}} \FunctionTok{ts\_mutate}\NormalTok{(ts, }\AttributeTok{mutation\_rate =} \DecValTok{1}\SpecialCharTok{*}\DecValTok{10}\SpecialCharTok{\^{}{-}}\DecValTok{4}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Now with mutations added and our populations defined, we can use the
\texttt{tskit} functions (all denoted in the package by starting with
\texttt{ts\_}) to calculate \emph{F}\textsubscript{ST}. This is extremly
fast and there are a large number of different functions to calculate
different statistics.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ts\_fst}\NormalTok{(ts\_m, }\AttributeTok{sample\_sets =} \FunctionTok{list}\NormalTok{(}\AttributeTok{a =}\NormalTok{ a\_pop, }\AttributeTok{b =}\NormalTok{ b\_pop))}
\end{Highlighting}
\end{Shaded}

We can also calculate diversity using the \texttt{ts\_diversity}
function - \textbf{nb} this is per site - to get \texttt{pi} for the
sequence, you would need to divide this value by the length (1000 in
this case).

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ts\_diversity}\NormalTok{(ts\_m, }\AttributeTok{sample\_sets =} \FunctionTok{list}\NormalTok{(}\AttributeTok{a =}\NormalTok{ a\_pop, }\AttributeTok{b =}\NormalTok{ b\_pop),}
             \AttributeTok{mode =} \StringTok{"site"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

So there we have it - two population genetic statistic estimates for a
simple 2-population model, all calculated in a single R framework at
high speed. If we placed all the commands we used together in an single
R script, we could have run this all in seconds, using very little hard
drive space.

\hypertarget{a-simple-simulation-pipeline}{%
\subsection{A simple simulation
pipeline}\label{a-simple-simulation-pipeline}}

Running a single simulation is useful for learning but it isn't that
helpful as a standalone tool. Instead, we can combine the code we've
learned to examine how the variaton in population parameters might
influence the statistics we calculate. Here we will run the model we
created above for different population sizes for population B (from 1000
to 15000) - how does it alter our estimate of
\emph{F}\textsubscript{ST}?

This code might look complex - but it is almost everything we covered
above!

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# set our sequence to simulate across}
\NormalTok{pop\_sizes }\OtherTok{\textless{}{-}} \FunctionTok{seq}\NormalTok{(}\DecValTok{1000}\NormalTok{, }\DecValTok{15000}\NormalTok{, }\AttributeTok{by =} \DecValTok{1000}\NormalTok{)}

\CommentTok{\# run our model within an sapply command}
\NormalTok{fst\_i }\OtherTok{\textless{}{-}} \FunctionTok{sapply}\NormalTok{(pop\_sizes, }\ControlFlowTok{function}\NormalTok{(x)\{}
  \CommentTok{\# set the sampling scheme}
\NormalTok{  samples }\OtherTok{\textless{}{-}} \FunctionTok{schedule\_sampling}\NormalTok{(model, }\AttributeTok{times =} \DecValTok{0}\NormalTok{, }\FunctionTok{list}\NormalTok{(a, }\DecValTok{30}\NormalTok{), }\FunctionTok{list}\NormalTok{(b, }\DecValTok{30}\NormalTok{))}
  \CommentTok{\# Population a}
\NormalTok{  a }\OtherTok{\textless{}{-}} \FunctionTok{population}\NormalTok{(}\StringTok{"a"}\NormalTok{, }\AttributeTok{time =} \DecValTok{50000}\NormalTok{, }\AttributeTok{N =} \DecValTok{30000}\NormalTok{)}
  \CommentTok{\# Population b }
\NormalTok{  b }\OtherTok{\textless{}{-}} \FunctionTok{population}\NormalTok{(}\StringTok{"b"}\NormalTok{, }\AttributeTok{parent =}\NormalTok{ a, }\AttributeTok{time =} \DecValTok{30000}\NormalTok{, }\AttributeTok{N =}\NormalTok{ x)}
  
  \CommentTok{\# compile model}
\NormalTok{  model }\OtherTok{\textless{}{-}} \FunctionTok{compile\_model}\NormalTok{(}
    \AttributeTok{populations =} \FunctionTok{list}\NormalTok{(a, b),}
    \AttributeTok{generation\_time =} \DecValTok{1}\NormalTok{,}
    \AttributeTok{direction =} \StringTok{"backward"}\NormalTok{,}
\NormalTok{  )}
  \CommentTok{\# run the model}
\NormalTok{  ts }\OtherTok{\textless{}{-}} \FunctionTok{msprime}\NormalTok{(model, }\AttributeTok{samples =}\NormalTok{ samples, }\AttributeTok{sequence\_length =} \DecValTok{1000}\NormalTok{, }\AttributeTok{recombination\_rate =} \DecValTok{0}\NormalTok{)}
  \CommentTok{\# add the mutations}
\NormalTok{  ts\_m }\OtherTok{\textless{}{-}} \FunctionTok{ts\_mutate}\NormalTok{(ts, }\AttributeTok{mutation\_rate =} \DecValTok{1}\SpecialCharTok{*}\DecValTok{10}\SpecialCharTok{\^{}{-}}\DecValTok{4}\NormalTok{)}
  \CommentTok{\# set the populations}
\NormalTok{  a\_pop }\OtherTok{\textless{}{-}} \FunctionTok{ts\_samples}\NormalTok{(ts) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{filter}\NormalTok{(pop }\SpecialCharTok{==}\NormalTok{ a) }\SpecialCharTok{\%\textgreater{}\%}\NormalTok{ .}\SpecialCharTok{$}\NormalTok{name}
\NormalTok{  b\_pop }\OtherTok{\textless{}{-}} \FunctionTok{ts\_samples}\NormalTok{(ts) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{filter}\NormalTok{(pop }\SpecialCharTok{==}\NormalTok{ b) }\SpecialCharTok{\%\textgreater{}\%}\NormalTok{ .}\SpecialCharTok{$}\NormalTok{name}
  \CommentTok{\# calculate fst}
\NormalTok{  y }\OtherTok{\textless{}{-}} \FunctionTok{ts\_fst}\NormalTok{(ts\_m, }\AttributeTok{sample\_sets =} \FunctionTok{list}\NormalTok{(}\AttributeTok{a =}\NormalTok{ a\_pop, }\AttributeTok{b =}\NormalTok{ b\_pop))}
\NormalTok{  y}\SpecialCharTok{$}\NormalTok{Fst}
\NormalTok{\})}
\end{Highlighting}
\end{Shaded}

We can then combine our results into a \texttt{data.frame} and plot the
results using \texttt{ggplot} to see how they vary across the
parameters.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# create a data.frame or tibble}
\NormalTok{sims }\OtherTok{\textless{}{-}} \FunctionTok{as\_tibble}\NormalTok{(}\FunctionTok{data.frame}\NormalTok{(}\AttributeTok{pop\_sizes =}\NormalTok{ pop\_sizes, }\AttributeTok{fst =}\NormalTok{ fst\_i))}

\CommentTok{\# plot the output}
\FunctionTok{ggplot}\NormalTok{(sims, }\FunctionTok{aes}\NormalTok{(pop\_sizes, fst)) }\SpecialCharTok{+} \FunctionTok{geom\_point}\NormalTok{() }\SpecialCharTok{+} \FunctionTok{geom\_line}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

Quite clearly from this simple simulation we can see that as the
populations size of B increases, so does the magnitude of
\emph{F}\textsubscript{ST} between the two populations.

\hypertarget{adding-to-the-model-gene-flow}{%
\subsubsection{Adding to the model: gene
flow}\label{adding-to-the-model-gene-flow}}

The model we have been using so far is very simple - it is basically a
2-deme isolation model. But what if we want to add gene flow between our
two populations? We can do this very easily using the
\texttt{gene\_flow} function.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gf }\OtherTok{\textless{}{-}}
  \FunctionTok{list}\NormalTok{(}\FunctionTok{gene\_flow}\NormalTok{(}\AttributeTok{from =}\NormalTok{ a, }\AttributeTok{to =}\NormalTok{ b, }\AttributeTok{rate =} \FloatTok{0.2}\NormalTok{, }\AttributeTok{start =} \DecValTok{10000}\NormalTok{, }\AttributeTok{end =} \DecValTok{0}\NormalTok{),}
       \FunctionTok{gene\_flow}\NormalTok{(}\AttributeTok{from =}\NormalTok{ b, }\AttributeTok{to =}\NormalTok{ a, }\AttributeTok{rate =} \FloatTok{0.2}\NormalTok{, }\AttributeTok{start =} \DecValTok{10000}\NormalTok{, }\AttributeTok{end =} \DecValTok{0}\NormalTok{)}
\NormalTok{  )}
\end{Highlighting}
\end{Shaded}

Here we have ensured that we are simulating gene flow that is symmetric
between our two populations at a rate of 0.2 between populations. This
starts 10,000 generations in the past and continues until the present
day. We have defined these two events in and combined them into a
\texttt{list} called \texttt{gf}

Next we compile our model, but this time include a gene flow term.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{model }\OtherTok{\textless{}{-}} \FunctionTok{compile\_model}\NormalTok{(}
  \AttributeTok{populations =} \FunctionTok{list}\NormalTok{(a, b),}
  \AttributeTok{gene\_flow =}\NormalTok{ gf,}
  \AttributeTok{generation\_time =} \DecValTok{1}\NormalTok{,}
  \AttributeTok{direction =} \StringTok{"backward"}\NormalTok{,}
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

And as before, we can plot the model to see what it looks like:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot\_model}\NormalTok{(model)}
\end{Highlighting}
\end{Shaded}

As before, we can run our model and see how it alters our estimates of
\emph{F}\textsubscript{ST} and diversity. We need to rerun our model,
populate it with mutations, define our populations and then calculate
the statistics.

\begin{Shaded}
\begin{Highlighting}[]
  \CommentTok{\# run the model}
\NormalTok{  ts }\OtherTok{\textless{}{-}} \FunctionTok{msprime}\NormalTok{(model, }\AttributeTok{samples =}\NormalTok{ samples, }\AttributeTok{sequence\_length =} \DecValTok{1000}\NormalTok{, }\AttributeTok{recombination\_rate =} \DecValTok{0}\NormalTok{)}
  \CommentTok{\# add the mutations}
\NormalTok{  ts\_m }\OtherTok{\textless{}{-}} \FunctionTok{ts\_mutate}\NormalTok{(ts, }\AttributeTok{mutation\_rate =} \DecValTok{1}\SpecialCharTok{*}\DecValTok{10}\SpecialCharTok{\^{}{-}}\DecValTok{4}\NormalTok{)}
  \CommentTok{\# set the populations}
\NormalTok{  a\_pop }\OtherTok{\textless{}{-}} \FunctionTok{ts\_samples}\NormalTok{(ts) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{filter}\NormalTok{(pop }\SpecialCharTok{==}\NormalTok{ a) }\SpecialCharTok{\%\textgreater{}\%}\NormalTok{ .}\SpecialCharTok{$}\NormalTok{name}
\NormalTok{  b\_pop }\OtherTok{\textless{}{-}} \FunctionTok{ts\_samples}\NormalTok{(ts) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{filter}\NormalTok{(pop }\SpecialCharTok{==}\NormalTok{ b) }\SpecialCharTok{\%\textgreater{}\%}\NormalTok{ .}\SpecialCharTok{$}\NormalTok{name}
  \CommentTok{\# calculate fst}
  \FunctionTok{ts\_fst}\NormalTok{(ts\_m, }\AttributeTok{sample\_sets =} \FunctionTok{list}\NormalTok{(}\AttributeTok{a =}\NormalTok{ a\_pop, }\AttributeTok{b =}\NormalTok{ b\_pop))}
  \CommentTok{\# calculate diversity}
  \FunctionTok{ts\_diversity}\NormalTok{(ts\_m, }\AttributeTok{sample\_sets =} \FunctionTok{list}\NormalTok{(}\AttributeTok{a =}\NormalTok{ a\_pop, }\AttributeTok{b =}\NormalTok{ b\_pop),}
             \AttributeTok{mode =} \StringTok{"site"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Yet again, a single value is interesting but it doesn't tell us too
much. So we will do what we did previously - rerunning our model but
altering the population size of population B. \textbf{However}, this
time we will see how gene flow influences our inference!

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# set our sequence to simulate across}
\NormalTok{pop\_sizes }\OtherTok{\textless{}{-}} \FunctionTok{seq}\NormalTok{(}\DecValTok{1000}\NormalTok{, }\DecValTok{15000}\NormalTok{, }\AttributeTok{by =} \DecValTok{1000}\NormalTok{)}

\CommentTok{\# run our model within an sapply command}
\NormalTok{fst\_g }\OtherTok{\textless{}{-}} \FunctionTok{sapply}\NormalTok{(pop\_sizes, }\ControlFlowTok{function}\NormalTok{(x)\{}
  \CommentTok{\# set the sampling scheme}
\NormalTok{  samples }\OtherTok{\textless{}{-}} \FunctionTok{schedule\_sampling}\NormalTok{(model, }\AttributeTok{times =} \DecValTok{0}\NormalTok{, }\FunctionTok{list}\NormalTok{(a, }\DecValTok{30}\NormalTok{), }\FunctionTok{list}\NormalTok{(b, }\DecValTok{30}\NormalTok{))}
  \CommentTok{\# Population a}
\NormalTok{  a }\OtherTok{\textless{}{-}} \FunctionTok{population}\NormalTok{(}\StringTok{"a"}\NormalTok{, }\AttributeTok{time =} \DecValTok{50000}\NormalTok{, }\AttributeTok{N =} \DecValTok{30000}\NormalTok{)}
  \CommentTok{\# Population b }
\NormalTok{  b }\OtherTok{\textless{}{-}} \FunctionTok{population}\NormalTok{(}\StringTok{"b"}\NormalTok{, }\AttributeTok{parent =}\NormalTok{ a, }\AttributeTok{time =} \DecValTok{30000}\NormalTok{, }\AttributeTok{N =}\NormalTok{ x)}
  
  \CommentTok{\# compile model {-} note the inclusion of gene flow}
\NormalTok{ model }\OtherTok{\textless{}{-}} \FunctionTok{compile\_model}\NormalTok{(}
  \AttributeTok{populations =} \FunctionTok{list}\NormalTok{(a, b),}
  \AttributeTok{gene\_flow =}\NormalTok{ gf,}
  \AttributeTok{generation\_time =} \DecValTok{1}\NormalTok{,}
  \AttributeTok{direction =} \StringTok{"backward"}\NormalTok{,}
\NormalTok{)}
 
  \CommentTok{\# run the model}
\NormalTok{  ts }\OtherTok{\textless{}{-}} \FunctionTok{msprime}\NormalTok{(model, }\AttributeTok{samples =}\NormalTok{ samples, }\AttributeTok{sequence\_length =} \DecValTok{1000}\NormalTok{, }\AttributeTok{recombination\_rate =} \DecValTok{0}\NormalTok{)}
  \CommentTok{\# add the mutations}
\NormalTok{  ts\_m }\OtherTok{\textless{}{-}} \FunctionTok{ts\_mutate}\NormalTok{(ts, }\AttributeTok{mutation\_rate =} \DecValTok{1}\SpecialCharTok{*}\DecValTok{10}\SpecialCharTok{\^{}{-}}\DecValTok{4}\NormalTok{)}
  \CommentTok{\# set the populations}
\NormalTok{  a\_pop }\OtherTok{\textless{}{-}} \FunctionTok{ts\_samples}\NormalTok{(ts) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{filter}\NormalTok{(pop }\SpecialCharTok{==}\NormalTok{ a) }\SpecialCharTok{\%\textgreater{}\%}\NormalTok{ .}\SpecialCharTok{$}\NormalTok{name}
\NormalTok{  b\_pop }\OtherTok{\textless{}{-}} \FunctionTok{ts\_samples}\NormalTok{(ts) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{filter}\NormalTok{(pop }\SpecialCharTok{==}\NormalTok{ b) }\SpecialCharTok{\%\textgreater{}\%}\NormalTok{ .}\SpecialCharTok{$}\NormalTok{name}
  \CommentTok{\# calculate fst}
\NormalTok{  y }\OtherTok{\textless{}{-}} \FunctionTok{ts\_fst}\NormalTok{(ts\_m, }\AttributeTok{sample\_sets =} \FunctionTok{list}\NormalTok{(}\AttributeTok{a =}\NormalTok{ a\_pop, }\AttributeTok{b =}\NormalTok{ b\_pop))}
\NormalTok{  y}\SpecialCharTok{$}\NormalTok{Fst}
\NormalTok{\})}
\end{Highlighting}
\end{Shaded}

Now we can add the output of the simulations from this run to our
previous simulations (those without gene flow) and see the difference

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# combine everything into a tibble}
\NormalTok{sims }\OtherTok{\textless{}{-}} \FunctionTok{as\_tibble}\NormalTok{(}\FunctionTok{data.frame}\NormalTok{(sims, fst\_g))}
\CommentTok{\# alter names! you will see why shortly }
\FunctionTok{colnames}\NormalTok{(sims) }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"pop\_sizes"}\NormalTok{, }\StringTok{"isolation"}\NormalTok{, }\StringTok{"gene\_flow"}\NormalTok{)}
\CommentTok{\# pivot to allow easy plotting}
\NormalTok{sims\_p }\OtherTok{\textless{}{-}} \FunctionTok{pivot\_longer}\NormalTok{(sims, }\SpecialCharTok{{-}}\NormalTok{pop\_sizes, }\AttributeTok{names\_to =} \StringTok{"model"}\NormalTok{, }\AttributeTok{values\_to =} \StringTok{"fst"}\NormalTok{)}
\CommentTok{\# plot the output}
\FunctionTok{ggplot}\NormalTok{(sims\_p, }\FunctionTok{aes}\NormalTok{(pop\_sizes, fst, }\AttributeTok{colour =}\NormalTok{ model)) }\SpecialCharTok{+} \FunctionTok{geom\_point}\NormalTok{() }\SpecialCharTok{+} \FunctionTok{geom\_line}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

This shows that changing the population size of b has the same result in
both models (i.e.~\emph{F}\textsubscript{ST} decreases with increasing
pop size) but that \emph{F}\textsubscript{ST} is lower in the gene flow
model - as we would expect!

\hypertarget{adding-to-the-model-resizing-a-population}{%
\subsubsection{Adding to the model: resizing a
population}\label{adding-to-the-model-resizing-a-population}}

As well as gene flow events, we can also resize populations so that they
experience bottlenecks or growth over time. Yet again, \texttt{slendr}
makes this very straightforward. All we need to do is pipe our
population declaration to a \texttt{resize} function when we declare
populations.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Population a with 30,000 individuals, arising 50000 generations ago}
\NormalTok{a }\OtherTok{\textless{}{-}} \FunctionTok{population}\NormalTok{(}\StringTok{"a"}\NormalTok{, }\AttributeTok{time =} \DecValTok{50000}\NormalTok{, }\AttributeTok{N =} \DecValTok{30000}\NormalTok{)}

\CommentTok{\# Population b with 15,000 individuals, arising 30,000 generations ago}
\NormalTok{b }\OtherTok{\textless{}{-}} \FunctionTok{population}\NormalTok{(}\StringTok{"b"}\NormalTok{, }\AttributeTok{parent =}\NormalTok{ a, }\AttributeTok{time =} \DecValTok{30000}\NormalTok{, }\AttributeTok{N =} \DecValTok{15000}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{resize}\NormalTok{(}\AttributeTok{N =} \DecValTok{2000}\NormalTok{, }\AttributeTok{how =} \StringTok{"step"}\NormalTok{, }\AttributeTok{time =} \DecValTok{5000}\NormalTok{, }\AttributeTok{end =} \DecValTok{0}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Remember because we are working with coalescent models, we are working
backwards in time. So here we set population b to start with a
population size of 2000 at time 0 and this then increases to the
ancestral population size 5000 generations in the past. Importantly we
set the \texttt{how} for this resize to \texttt{step} - i.e.~it will
just change suddenly.

Then we need to just declare our model again. For simplicity here, we'll
do this without gene flow.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{model }\OtherTok{\textless{}{-}} \FunctionTok{compile\_model}\NormalTok{(}
  \AttributeTok{populations =} \FunctionTok{list}\NormalTok{(a, b),}
  \AttributeTok{generation\_time =} \DecValTok{1}\NormalTok{,}
  \AttributeTok{direction =} \StringTok{"backward"}\NormalTok{,}
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Then we can plot it to see how this looks!

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot\_model}\NormalTok{(model)}
\end{Highlighting}
\end{Shaded}

And what if we had set this to happen exponentially, rather than a
sudden step?

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Population a with 30,000 individuals, arising 50000 generations ago}
\NormalTok{a }\OtherTok{\textless{}{-}} \FunctionTok{population}\NormalTok{(}\StringTok{"a"}\NormalTok{, }\AttributeTok{time =} \DecValTok{50000}\NormalTok{, }\AttributeTok{N =} \DecValTok{30000}\NormalTok{)}

\CommentTok{\# Population b with 15,000 individuals, arising 30,000 generations ago}
\NormalTok{b }\OtherTok{\textless{}{-}} \FunctionTok{population}\NormalTok{(}\StringTok{"b"}\NormalTok{, }\AttributeTok{parent =}\NormalTok{ a, }\AttributeTok{time =} \DecValTok{30000}\NormalTok{, }\AttributeTok{N =} \DecValTok{15000}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{resize}\NormalTok{(}\AttributeTok{N =} \DecValTok{2000}\NormalTok{, }\AttributeTok{how =} \StringTok{"exponential"}\NormalTok{, }\AttributeTok{time =} \DecValTok{5000}\NormalTok{, }\AttributeTok{end =} \DecValTok{0}\NormalTok{)}

\CommentTok{\# recompile the model}
\NormalTok{model }\OtherTok{\textless{}{-}} \FunctionTok{compile\_model}\NormalTok{(}
  \AttributeTok{populations =} \FunctionTok{list}\NormalTok{(a, b),}
  \AttributeTok{generation\_time =} \DecValTok{1}\NormalTok{,}
  \AttributeTok{direction =} \StringTok{"backward"}\NormalTok{,}
\NormalTok{)}

\CommentTok{\# plot the model}
\FunctionTok{plot\_model}\NormalTok{(model)}
\end{Highlighting}
\end{Shaded}

We won't include the resizing in our simulation pipeline above because
next we will try to develop a simple spatial simulation - this will be a
very useful tool for landscape genomics!

\hypertarget{spatial-models}{%
\subsection{Spatial models}\label{spatial-models}}

So far, our models have all been relatively basic with no spatial
context. But what if adding a spatial context helped make them more
realistic? This is a very difficult and complex topic but it is the main
motivation behind the development of \texttt{slendr} - again another
reason to refer to \href{https://www.slendr.net/}{its excellent and
extensive website}.

\hypertarget{setting-up-the-spatial-context}{%
\subsubsection{Setting up the spatial
context}\label{setting-up-the-spatial-context}}

We will try to take our basic model and put it in a spatial context in
order to get a flavour of the kind of things you can do with
\texttt{slendr}. The very first thing we will do is define a map that we
will use - we use the \texttt{world} function to do this - and we simply
set the longitude (\texttt{xrange}) and latitude (\texttt{yrange}) to do
this.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{map }\OtherTok{\textless{}{-}} \FunctionTok{world}\NormalTok{(}
  \AttributeTok{xrange =} \FunctionTok{c}\NormalTok{(}\SpecialCharTok{{-}}\DecValTok{13}\NormalTok{, }\DecValTok{70}\NormalTok{), }\CommentTok{\# min{-}max longitude}
  \AttributeTok{yrange =} \FunctionTok{c}\NormalTok{(}\DecValTok{18}\NormalTok{, }\DecValTok{65}\NormalTok{),  }\CommentTok{\# min{-}max latitude}
  \AttributeTok{crs =} \StringTok{"EPSG:3035"}    \CommentTok{\# coordinate reference system (CRS) for West Eurasia}
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

With this set, we can then plot the map using the \texttt{plot\_map}
function from \texttt{slendr} to make an easy map as a background for
simulations.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot\_map}\NormalTok{(map)}
\end{Highlighting}
\end{Shaded}

Next we will create two regions on our map - one over the UK, the other
over Europe.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# anatolia}
\NormalTok{anatolia }\OtherTok{\textless{}{-}} \FunctionTok{region}\NormalTok{(}
  \StringTok{"Anatolia"}\NormalTok{, map,}
  \AttributeTok{polygon =} \FunctionTok{list}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\DecValTok{28}\NormalTok{, }\DecValTok{35}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{40}\NormalTok{, }\DecValTok{35}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{42}\NormalTok{, }\DecValTok{40}\NormalTok{),}
                 \FunctionTok{c}\NormalTok{(}\DecValTok{30}\NormalTok{, }\DecValTok{43}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{27}\NormalTok{, }\DecValTok{40}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{25}\NormalTok{, }\DecValTok{38}\NormalTok{))}
\NormalTok{)}
\CommentTok{\# europe}
\NormalTok{europe }\OtherTok{\textless{}{-}} \FunctionTok{region}\NormalTok{(}
  \StringTok{"Europe"}\NormalTok{, map,}
  \AttributeTok{polygon =} \FunctionTok{list}\NormalTok{(}
    \FunctionTok{c}\NormalTok{(}\SpecialCharTok{{-}}\DecValTok{10}\NormalTok{, }\DecValTok{35}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\SpecialCharTok{{-}}\DecValTok{5}\NormalTok{, }\DecValTok{36}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{10}\NormalTok{, }\DecValTok{38}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{20}\NormalTok{, }\DecValTok{35}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{23}\NormalTok{, }\DecValTok{35}\NormalTok{),}
    \FunctionTok{c}\NormalTok{(}\DecValTok{30}\NormalTok{, }\DecValTok{45}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{20}\NormalTok{, }\DecValTok{52}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{50}\NormalTok{), }\FunctionTok{c}\NormalTok{(}\SpecialCharTok{{-}}\DecValTok{10}\NormalTok{, }\DecValTok{48}\NormalTok{)}
\NormalTok{  )}
\NormalTok{)}
\FunctionTok{plot\_map}\NormalTok{(anatolia, europe)}
\end{Highlighting}
\end{Shaded}

So here we have a map with two polygons imposed on the top that define
regions. With this spatial structure set up, we can now start to build a
model that is anchored in this geographical context.

\hypertarget{building-a-spatial-model}{%
\subsubsection{Building a spatial
model}\label{building-a-spatial-model}}

As with our simpler, non-spatial models, we can start to specify our
model using the same functions as previously - i.e.~\texttt{population},
except this time we actually incorporate the map data.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# european population}
\NormalTok{eur }\OtherTok{\textless{}{-}} \FunctionTok{population}\NormalTok{(}
  \AttributeTok{name =} \StringTok{"eur"}\NormalTok{, }\AttributeTok{time =} \DecValTok{6000}\NormalTok{, }\AttributeTok{N =} \DecValTok{2000}\NormalTok{,}
  \AttributeTok{polygon =}\NormalTok{ europe, }\AttributeTok{map =}\NormalTok{ map}
\NormalTok{)}
\CommentTok{\# check it by plotting!}
\FunctionTok{plot\_map}\NormalTok{(eur)}
\CommentTok{\# anatolian population}
\NormalTok{ana }\OtherTok{\textless{}{-}} \FunctionTok{population}\NormalTok{( }\CommentTok{\# Anatolian pop}
  \AttributeTok{name =} \StringTok{"ana"}\NormalTok{, }\AttributeTok{time =} \DecValTok{6000}\NormalTok{, }\AttributeTok{N =} \DecValTok{3000}\NormalTok{,}
  \AttributeTok{center =} \FunctionTok{c}\NormalTok{(}\DecValTok{34}\NormalTok{, }\DecValTok{38}\NormalTok{), }\AttributeTok{radius =} \FloatTok{500e3}\NormalTok{, }\AttributeTok{polygon =}\NormalTok{ anatolia, }\AttributeTok{map =}\NormalTok{ map}
\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{expand\_range}\NormalTok{( }\CommentTok{\# expand the range by 2.500 km}
    \AttributeTok{by =} \FloatTok{2500e3}\NormalTok{, }\AttributeTok{start =} \DecValTok{5000}\NormalTok{, }\AttributeTok{end =} \DecValTok{3000}\NormalTok{, }\AttributeTok{overlap =} \FloatTok{0.5}\NormalTok{,}
    \AttributeTok{polygon =} \FunctionTok{join}\NormalTok{(europe, anatolia)}
\NormalTok{  )}
\end{Highlighting}
\end{Shaded}

Note that the arguments \texttt{map} and \texttt{polygon} allow us to
specify the map and polygon we have already defined - it is these
arguments which give our population its spatial rooting.

With this done, we can now replot these polygons and you will see we
have defined the ranges of the populations within the polygons - here
they are explicitly bounded to the landscape. We will see why this is
important shortly.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot\_map}\NormalTok{(ana)}
\CommentTok{\# plot both together}
\FunctionTok{plot\_map}\NormalTok{(eur, ana) }\CommentTok{\# showing an expansion into europe}
\end{Highlighting}
\end{Shaded}

With our populations defined, we can also set out some gene flow events.
We will do this exactly the same way we did with our non-spatial model
earlier.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# this will not work}
\NormalTok{gf }\OtherTok{\textless{}{-}} \FunctionTok{gene\_flow}\NormalTok{(}\AttributeTok{from =}\NormalTok{ ana, }\AttributeTok{to =}\NormalTok{ eur, }\AttributeTok{rate =} \FloatTok{0.1}\NormalTok{, }\AttributeTok{start =} \DecValTok{5000}\NormalTok{, }\AttributeTok{end =} \DecValTok{4000}\NormalTok{, }\AttributeTok{overlap =}\NormalTok{ T)}
\CommentTok{\# this will }
\NormalTok{gf }\OtherTok{\textless{}{-}} \FunctionTok{gene\_flow}\NormalTok{(}\AttributeTok{from =}\NormalTok{ ana, }\AttributeTok{to =}\NormalTok{ eur, }\AttributeTok{rate =} \FloatTok{0.1}\NormalTok{, }\AttributeTok{start =} \DecValTok{3000}\NormalTok{, }\AttributeTok{end =} \DecValTok{2000}\NormalTok{, }\AttributeTok{overlap =}\NormalTok{ T)}
\end{Highlighting}
\end{Shaded}

The main difference here however is that we have now added an
\texttt{overlap} argument. This is basically a requirement that
populations must spatially overlap in order to to exchange genes.

With this done, we can then compile our model. The principle here is the
same as with our non-spatial model but with some additional arguments.
We will learn about these after we have run the command. Also, ensure
population names are correct!

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# compile model}
\NormalTok{model\_dir }\OtherTok{\textless{}{-}} \FunctionTok{paste0}\NormalTok{(}\FunctionTok{tempfile}\NormalTok{(), }\StringTok{"\_tutorial{-}model"}\NormalTok{)}


\NormalTok{model }\OtherTok{\textless{}{-}} \FunctionTok{compile\_model}\NormalTok{(}
  \AttributeTok{populations =} \FunctionTok{list}\NormalTok{(eur, ana), }\CommentTok{\# populations defined above}
  \AttributeTok{gene\_flow =}\NormalTok{ gf, }\CommentTok{\# gene{-}flow events defined above}
  \AttributeTok{generation\_time =} \DecValTok{30}\NormalTok{,}
  \AttributeTok{resolution =} \FloatTok{100e3}\NormalTok{, }\CommentTok{\# resolution in meters per pixel}
  \AttributeTok{competition =} \FloatTok{130e3}\NormalTok{, }\AttributeTok{mating =} \FloatTok{100e3}\NormalTok{, }\CommentTok{\# spatial interaction in SLiM}
  \AttributeTok{dispersal =} \FloatTok{700e3}\NormalTok{, }\CommentTok{\# how far will offspring end up from their parents}
  \AttributeTok{path =}\NormalTok{ model\_dir}
\NormalTok{)}

\FunctionTok{plot\_model}\NormalTok{(model, }\AttributeTok{proportions =}\NormalTok{ T) }\CommentTok{\# to check}
\end{Highlighting}
\end{Shaded}

So what additional arguments have we added here that differs from our
previous model compliation in the non-spatial examples?

\begin{itemize}
\tightlist
\item
  Firstly we have the \texttt{resolution} argument. This is simply the
  resolution of the map to simulate on and how much a single pixel
  represents. Here we have set it to 100,000 units.
\item
  Next we have \texttt{competition} - this is the maximum distance
  between two individuals where they can influence each others fitness
  via competition. Here it is set to 130,000 which basically menas
  individuals influence each other only if the occur right next to one
  another on the map.
\item
  We also have \texttt{mating} - this sets the mating choice distance,
  i.e.~the maximum distance an individual will find a mate over; set to
  100,000 here, it means individuals look for mates in close proximity.
\item
  Last we have \texttt{dispersal} which is fairly self-explanatory as
  dispersal distance. In the context of our model, it determines how far
  an individual can move before contributing to next generation.
\end{itemize}

Note that we also specify a \texttt{path} to set a model directory, just
so we can look at the files SLiM writes to the directory should we need
to.

Finally we use \texttt{plot\_model} to make sure the model is doing what
we expect. If we are satisfied with this, we can now run it using
\texttt{slim}. Note that this is a large difference from our non-spatial
models which used \texttt{msprime}. SliM is a forward in time simulator
which allows it to incorporate selection and spatial dynamics.
\href{https://messerlab.org/slim/}{It is extremely powerful} and I would
strongly recommend you investigate it in more detail!

However one disadvantage of SLiM is that it is slower than
\texttt{msprime} - this menas our simulation will take a bit longer to
complete\ldots{} but not too long!

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# set locations file}
\NormalTok{locations\_file }\OtherTok{\textless{}{-}} \FunctionTok{tempfile}\NormalTok{(}\AttributeTok{fileext =} \StringTok{".gz"}\NormalTok{)}
\CommentTok{\# run the simulations}
\NormalTok{ts }\OtherTok{\textless{}{-}} \FunctionTok{slim}\NormalTok{(model, }\AttributeTok{sequence\_length =} \DecValTok{1000}\NormalTok{, }\AttributeTok{recombination\_rate =} \DecValTok{0}\NormalTok{, }\AttributeTok{method =} \StringTok{"batch"}\NormalTok{,}
           \AttributeTok{locations =}\NormalTok{ locations\_file)}
\CommentTok{\# look at the tree sequence}
\NormalTok{ts}
\end{Highlighting}
\end{Shaded}

And with that, we're done. Hopefully this has given you a taste of what
is possible with \texttt{slendr}. There is so much more you could do and
many ways to extend and expand these models. It is well worth spending
some time with this excellent R package!

\hypertarget{optional-extra---animating-your-model}{%
\subsubsection{Optional extra - animating your
model}\label{optional-extra---animating-your-model}}

This actually doesn't work on our RStudio Server \textbf{but} it should
work locally. This is just an optional example that allows you to see
some of the ways \texttt{slendr} allows you to explore your simulations
and models.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# animate the model}
\FunctionTok{animate\_model}\NormalTok{(}\AttributeTok{model =}\NormalTok{ model, }\AttributeTok{file =}\NormalTok{ locations\_file, }\AttributeTok{steps =} \DecValTok{50}\NormalTok{, }\AttributeTok{width =} \DecValTok{700}\NormalTok{, }\AttributeTok{height =} \DecValTok{400}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{fastsimcoal}{%
\section{fastsimcoal}\label{fastsimcoal}}

Firstly before proceeding I want make it clear that this tutorial was
originally designed by my colleague and friend,
\href{https://www.sanger.ac.uk/person/meier-joana/}{Joana Meier} and has
been adapted from
\href{https://speciationgenomics.github.io/fastsimcoal2/}{here} for this
course. The depth of this tutorial would not have been possible without
her input!

\texttt{fastsimcoal2} is an extremely flexible demographic modelling
software developed by
\href{http://www.cmpg.iee.unibe.ch/about_us/team/researchers/prof_dr_excoffier_laurent/index_eng.html}{Laurent
Excoffier and his group at University of Bern}. It uses the site
frequency spectrum (SFS) to fit model parameters to the observed data by
performing coalescent simulations. You can find the manual and more
information \href{http://cmpg.unibe.ch/software/fastsimcoal2/}{here}.

It is worth noting that this is a complex piece of software to run and
it takes a lot of time to master. However it is really worth taking your
time with it and exploring the manual. It is extremely well documented
(which is rare for a program like this).

\hypertarget{preparing-the-input-files}{%
\subsubsection{Preparing the input
files}\label{preparing-the-input-files}}

To run \texttt{fastsimcoal2} we need three input files which are all
named in a consistent way. They are all just plain text files.

\begin{itemize}
\tightlist
\item
  observed SFS - \texttt{\$\{PREFIX\}\_jointDAFpop1\_0.obs}
\item
  template file defining the demographic model -
  \texttt{\$\{PREFIX\}.tpl}
\item
  estimation file defining the parameters - \texttt{\$\{PREFIX\}.est}
\end{itemize}

\hypertarget{observed-sfs}{%
\subsubsection{Observed SFS}\label{observed-sfs}}

The observed SFS can be a derived SFS (i.e.~also known as DAF or an
unfolded SFS) if the ancestral state is unknown or a minor allele
frequency SFS (i.e.~MAF or folded SFS) if you do know the ancestral
state. \texttt{fastsimcoal2} will identify the kind of SFS by its name
suffix and the command flag \texttt{-m} or \texttt{-d}. For a single
population, it expects the name \texttt{\$\{PREFIX\}\_DAFpop0.obs} or
\texttt{\$\{PREFIX\}\_MAFpop0.obs}.

For two populations, the file names should end with
\texttt{jointDAFpop1\_0.obs} or \texttt{jointMAFpop1\_0.obs}. If more
than two observed populations are used, pairwise MAFs or DAFs can be
provided following the naming scheme above or a multidimensional SFS can
be used ending with \texttt{\_DSFS.obs} or \texttt{\_MSFS.obs} and
specifying the \texttt{-multiSFS} flag when running the program. See the
\href{http://cmpg.unibe.ch/software/fastsimcoal2/}{fastsimcoal2 manual}
for further details. The \texttt{.obs} file can be generated with
\texttt{Arlequin}, \texttt{angsd}, \texttt{easySFS} (as in this
\href{https://speciationgenomics.github.io/easysfs/}{tutorial}) or
several other tools.

\hypertarget{template-file}{%
\subsubsection{Template file}\label{template-file}}

The template file describes the demographic model and the parameters of
interest. Values that should be estimated (i.e.~model parameters such as
migration rate, splitting time etc) are given as different keywords. It
is best practise to use capital letters for parameter names and it is
important to avoid any names that are part of another parameter name or
a function (e.g.~\texttt{log}, \texttt{min}, \texttt{FREQ}).

As fastsimcoal2 is a coalescent simulator, all models need to be
specified \textbf{backwards in time}. This means that the most recent
event is specified first. The template file is best produced by
modifying an example \texttt{tpl} file in a text editor.

In this tutorial, we will first write a model of two species that
diverged in the face of gene flow and then evolved complete reproductive
isolation. Unfortunately, we do not have a good estimate for the
mutation rate, but we have an estimate for the (maximum possible)
splitting time. So here we will set the split between the populations to
6000 generations.

\textbf{Question:} Why do we need to fix a parameter (e.g.~splitting
time) if we do not have a reliable mutation rate estimate?

\hypertarget{estimation-file}{%
\subsubsection{Estimation file}\label{estimation-file}}

All keywords introduced in the template file need to be defined in the
estimation file. For each keyword, the parameter distribution (uniform
or log-uniform) and search range (min and max) are given on a single
line. Each parameter can be an integer or a float, as specified by a
first indicator variable.

Lastly, complex parameters can be used to compute parameter values with
simple operations such as computing the ratio between two simple
parameters or specifying a parameter as the minimum of two parameters.
In addition, for each simple or complex parameter, you need to specify
if the parameter value should be written into an output file or not.

As with the template file, the estimation file is best produced by
modifying an example \texttt{est} file in a text editor.

\hypertarget{running-fastsimcoal2}{%
\subsection{Running fastsimcoal2}\label{running-fastsimcoal2}}

Once we have all input files ready, it is time to run
\texttt{fastsimcoal2}. In addition to the input files, we need to
specify how many simulations and iterations \texttt{fastsimcoal2} should
perform, when to stop and how many threads (i.e.~CPUs) can be used in
parallel.

Let's run \texttt{fastsimcoal2} with our model of two populations. First
we make a \texttt{fastsimcoal2} directory and within that a directory,
an additional one for our model of \texttt{early\_geneflow}:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cd \textasciitilde{}}
\NormalTok{mkdir fastsimcoal2}
\NormalTok{cd fastsimcoal2}
\NormalTok{mkdir early\_geneflow}
\end{Highlighting}
\end{Shaded}

Next, we will move inside this directory and copy over the files we need

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cd \textasciitilde{}/fastsimcoal2/early\_geneflow}
\NormalTok{cp /resources/riverlandsea/exercise\_data/fastsimcoal2/early\_geneflow/* .}
\end{Highlighting}
\end{Shaded}

This should mean we have our observed SFS, our template file and our
estimation file. We are now ready to run \texttt{fastsimcoal2} - we will
set a \texttt{PREFIX} variable to make our command here a bit easier to
write. This will help downstream too!

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{PREFIX="early\_geneflow"}
\NormalTok{fastsimcoal2 {-}t $\{PREFIX\}.tpl {-}e $\{PREFIX\}.est {-}m {-}0 {-}C 10 {-}n 10000 {-}L 40 {-}s 0 {-}M}
\end{Highlighting}
\end{Shaded}

This command runs \texttt{fastsimcoal2} using a MAF (\texttt{-m}) while
ignoring monomorphic sites (\texttt{-0}) and SFS entries with less than
10 SNPs (\texttt{-C}). This means that entries with less than 10 SNPs
are pooled together. This option is useful when there are many entries
in the observed SFS with few SNPs and with a limited number of SNPS to
avoid overfitting.

\texttt{fastsimcoal2} will also perform (\texttt{-n}) 10,000 coalescent
simulations to approximate the expected SFS in each cycle and will run
(\texttt{-L}) 40 optimization (ECM) cycles to estimate the parameters.
The number of ECM cycles should be at least 20, better between 50 and
100. The number of coalescent simulations should ideally be something
between 200,000 and 1,000,000 but to make it faster, we are now only
running 10,000 simulations. We also specify (\texttt{-M}) that we want
to perform parameter estimation. With \texttt{-s\ 0}, we can tell
\texttt{fastsimcoal2} to output SNPs.

Once \texttt{fastsimcoal2} is finished, we can have a look at the output
files. It produced a folder called \texttt{\$\{PREFIX\}} which contains
a number of new files. The most relevant files are the following:

\begin{itemize}
\tightlist
\item
  \texttt{\$\{PREFIX\}.bestlhoods:} a file with the Maximum likelihood
  estimates for each parameter specified ``output'' in the \texttt{est}
  file and the model likelihoods.
\item
  \texttt{\$\{PREFIX\}.\_jointMAFpop1\_0.txt}: a file with the expected
  SFS obtained with the parameters that maximized the likelihood during
  optimization. This is needed to visually check the fit of the expected
  SFS to the observed SFS. This file has the same suffix as the observed
  SFS provided.
\item
  \texttt{\$\{PREFIX\}.simparam}: a file with an example of the settings
  to run the simulations. This is useful to check when you have errors.
  Many times errors in specification of models can be detected in this
  file.
\item
  \texttt{\$\{PREFIX\}\_maxL.par}: The model specification file with the
  best parameter estimates. It is basically the tpl file with the
  keywords replaced by estimated values. This file is useful if you want
  to simulate data under the best model using Arlequin.
\end{itemize}

Note, the bestlhoods file contains two different likelihoods:
\texttt{MaxObsLhood} is the maximum possible value for the likelihood if
there was a perfect fit of the expected to the observed SFS, i.e.~if the
expected SFS was the relative observed SFS. \texttt{MaxEstLhood} is the
maximum likelihood estimated according to the model parameters. It is
obtained by using the observed SFS as the expected SFS when computing
the likelihood, i.e., returning the value of the likelihood if there was
a perfect fit between the expected and observed SFS.

The better the fit, the smaller the difference between
\texttt{MaxObsLhood} and \texttt{MaxEstLhood}.

\hypertarget{finding-the-best-parameter-estimates}{%
\subsubsection{Finding the best parameter
estimates}\label{finding-the-best-parameter-estimates}}

\texttt{fastsimcoal2} should not just be run once because it might not
find the global optimum of the best combination of parameter estimates
right away. It is better to run it 100 times or more. Of these runs, we
then go on to select the one with the highest likelihood which is the
run with the best fitting parameter estimates for this model.

Due to time constraints, we will only run this model 5 times and we will
do so within a \texttt{for} loop. Note, that here we have added the flag
\texttt{-q} for ``quiet'' which reduces the amount of information
\texttt{fastsimcoal2} writes to stdout. Pay attention the \texttt{cd}
commands here as they ensure that the analyses are done in the correct
directory.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ for i in \{1..5\}}
\NormalTok{ do}
\NormalTok{   mkdir run$i}
\NormalTok{   cp $\{PREFIX\}.tpl $\{PREFIX\}.est $\{PREFIX\}\_jointMAFpop1\_0.obs run$i"/"}
\NormalTok{   cd run$i}
\NormalTok{   fastsimcoal2 {-}t $\{PREFIX\}.tpl {-}e $\{PREFIX\}.est {-}m {-}0 {-}C 10 {-}n 100 {-}L 40 {-}s0 {-}M {-}q}
\NormalTok{   cd ..}
\NormalTok{ done}
\end{Highlighting}
\end{Shaded}

To find the best run, i.e.~the run with the highest likelihood, or
better the smallest difference between the maximum possible likelihood
(\texttt{MaxObsLhood}) and the obtained likelihood
(\texttt{MaxEstLhood}), we can check the \texttt{.bestlhoods} files.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cat run\{1..5\}/$\{PREFIX\}/$\{PREFIX\}.bestlhoods | grep {-}v MaxObsLhood | awk \textquotesingle{}\{print NR,$8\}\textquotesingle{} | sort {-}k 2}
\end{Highlighting}
\end{Shaded}

Note that \texttt{NR} in awk prints out the line number which here
corresponds to the run number. \texttt{\$8} is the \texttt{MaxEstLhood}
column and thus the likelihood we want to compare across different runs.

{[}Joana Meier{]}(https://www.sanger.ac.uk/person/meier-joana/ has
written a
\href{https://github.com/speciationgenomics/scripts/raw/master/fsc-selectbestrun.sh}{script}
for you that automatically extracts the files of the best run and copies
them into a new folder which it calls \texttt{bestrun}. A copy of that
script is available in the course resources directory so we will copy it
to our run folder directory and run it like so:

Just run it in the directory where all the folders run are located and
run it:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cd \textasciitilde{}/fastsimcoal2/early\_geneflow}
\NormalTok{cp /resources/riverlandsea/exercise\_data/fastsimcoal2/fsc{-}selectbestrun.sh .}
\NormalTok{./fsc{-}selectbestrun.sh}
\end{Highlighting}
\end{Shaded}

Now modify the \texttt{\$PREFIX.tpl} and \texttt{\$PREFIX.est} files to
specify different models and also rename the SFS to
\texttt{\$\{PREFIX\}\_jointDAFpop1\_0.txt}. Then run
\texttt{fastsimcoal2}for all models to see which one shows the best fit
to the observed SFS.

\includegraphics{./images/modeling/models.png}

Click here to see a possible solution.

Early Geneflow (this is the model expalined in the example, gene flow
right after the split and then no gene flow anymore, e.g.~speciation
with gene flow and then no gene flow anymore as reproductive isolation
becomes strong): tpl est No geneflow: tpl est Recent geneflow (no gene
flow intially after the split but gene flow in recent times,
e.g.~secondary contact scenario) tpl est Different gene flow matrices
(higher or lower gene flow right after the split than recently,
e.g.~initially higher gene flow then lower gene flow as reproductive
isolation accumulates): tpl est Constant gene flow (same gene flow
strengths since the split until now): tpl est

\hypertarget{model-comparison-with-aic}{%
\subsubsection{Model comparison with
AIC}\label{model-comparison-with-aic}}

In order to find the best model, the likelihoods of the best run of each
model should be compared. Comparing raw likelihoods is problematic,
because a model with more parameters will always tend to result in a
better fit to the data. Therefore, the
\href{https://en.wikipedia.org/wiki/Akaike_information_criterion}{Akaike
information criterium} or AIC is typically calculated to determine if
the models differ in their likelihoods accounting for the number of
parameters in each model. To perform, this, we can use a
\href{https://github.com/speciationgenomics/scripts/blob/master/calculateAIC.sh}{script}
that is mostly based on R code by
\href{http://ce3c.ciencias.ulisboa.pt/member/vitorsousa}{Vitor Sousa}.
We will move into our \texttt{bestrun} directory and run this script

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cd bestrun/}
\NormalTok{cp /resources/riverlandsea/exercise\_data/fastsimcoal2/calculateAIC.sh .}
\NormalTok{./calculateAIC.sh early\_geneflow}
\end{Highlighting}
\end{Shaded}

This script generates a file \texttt{\$\{PREFIX\}.AIC} which contains
the delta likelihood and the AIC value for that run.

\hypertarget{visualize-the-model-fit}{%
\subsubsection{Visualize the model fit}\label{visualize-the-model-fit}}

To visualize the fit of the simulated SFS to the data, we can use an
\texttt{R} script that David Marques wrote - \texttt{SFStools.r} - which
you can download \href{https://github.com/marqueda/SFS-scripts/}{here}.

To visualize the model with the best parameter estimates, we can use one
of Joana Meier's R scripts -
\href{https://github.com/speciationgenomics/scripts/blob/master/plotModel.r}{plotModel.r}.
There are also other options
\href{http://cmpg.unibe.ch/software/fastsimcoal27/additionalScripts.html}{here}.
However for the ease of this practical, both are available in the
resources directory. We will copy them over and then run them.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cp /resources/riverlandsea/exercise\_data/fastsimcoal2/*.r .}
\NormalTok{Rscript SFStools.r {-}t print2D {-}i early\_geneflow}
\NormalTok{Rscript plotModel.r {-}p early\_geneflow {-}l NyerMak,PundMak}
\end{Highlighting}
\end{Shaded}

Now, we will download the PDFs genertated and take a look at them.

\hypertarget{model-comparison-with-likelihood-distributions}{%
\subsubsection{Model comparison with Likelihood
distributions}\label{model-comparison-with-likelihood-distributions}}

One drawback of the composite likelihoods in model tests based on AIC is
that it can overestimate the support for the most likely model if the
SNPs are not independent (here they are not LD-pruned). Another way to
infer if the models are really different and do not just differ because
of stochasticity in the likelihood approximation, is to get likelihood
distributions for each model. This is done by running each model with
the best parameter values multple times (ideally about 100 times). The
likelihoods will differ because \texttt{fastsimcoal2} does not compute
the likelihood but rather approximates it with simulations. If the
ranges of likelihoods of two models overlap, it means that they do not
differ significantly, i.e.~provide an equally good fit to the observed
data.

Let's recompute the likelihood for the best run for the
\texttt{early\_geneflow} model.

\begin{Shaded}
\begin{Highlighting}[]

\NormalTok{PREFIX="early\_geneflow"}
\NormalTok{cd \textasciitilde{}/fastsimcoal/$PREFIX/bestrun}

\NormalTok{\# create temporary obs file with name \_maxL\_MSFS.obs}
\NormalTok{cp $\{PREFIX\}\_jointMAFpop1\_0.obs $\{PREFIX\}\_maxL\_jointMAFpop1\_0.obs}

\NormalTok{\# Run fastsimcoal 20 times (in reality better 100 times) to get the likelihood of the observed SFS under the best parameter values with 1 mio simulated SFS.}
\NormalTok{for i in \{1..20\}}
\NormalTok{do}
\NormalTok{ fastsimcoal2 {-}i $\{PREFIX\}\_maxL.par {-}n1000000 {-}m {-}q {-}0}
\NormalTok{ \# Fastsimcoal will generate a new folder called $\{model\}\_maxL and write files in there}

\NormalTok{ \# collect the lhood values (Note that \textgreater{}\textgreater{} appends to the file, whereas \textgreater{} would overwrite it)}
\NormalTok{ sed {-}n \textquotesingle{}2,3p\textquotesingle{} $\{PREFIX\}\_maxL/$\{PREFIX\}\_maxL.lhoods  \textgreater{}\textgreater{} $\{PREFIX\}.lhoods}

\NormalTok{ \# delete the folder with results}
\NormalTok{ rm {-}r $\{PREFIX\}\_maxL/}
\NormalTok{done}
\end{Highlighting}
\end{Shaded}

We would now repeat this for different models: ongoing gene flow, a
model without gene flow, a model of secondary contact (recent gene flow)
and a model with different amounts of gene flow in the past and in
recent times.

Due to time constraints, we will give you the results of these models.
They are in the same format as the folder we generated for the model
with early gene flow. Load them into your own directory. Note, the
\texttt{-r} flag stands for ``recursive'' and allows to also copy
directories and their contents.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cd \textasciitilde{}/fastsimcoal2/}
\NormalTok{cp {-}r /resources/riverlandsea/exercise\_data/fastsimcoal2/extramodels}
\end{Highlighting}
\end{Shaded}

Now, let's access the R Studio sderver and plot the likelihoods.

\begin{Shaded}
\begin{Highlighting}[]


\CommentTok{\# Read in the likelihoods}
\NormalTok{early\_geneflow}\OtherTok{\textless{}{-}}\FunctionTok{scan}\NormalTok{(}\StringTok{"early\_geneflow.lhoods"}\NormalTok{)}
\NormalTok{ongoing\_geneflow}\OtherTok{\textless{}{-}}\FunctionTok{scan}\NormalTok{(}\StringTok{"ongoing\_geneflow.lhoods"}\NormalTok{)}
\NormalTok{diff\_geneflow}\OtherTok{\textless{}{-}}\FunctionTok{scan}\NormalTok{(}\StringTok{"diff\_geneflow.lhoods"}\NormalTok{)}
\NormalTok{recent\_geneflow}\OtherTok{\textless{}{-}}\FunctionTok{scan}\NormalTok{(}\StringTok{"recent\_geneflow.lhoods"}\NormalTok{)}
\NormalTok{no\_geneflow}\OtherTok{\textless{}{-}}\FunctionTok{scan}\NormalTok{(}\StringTok{"no\_geneflow.lhoods"}\NormalTok{)}

\CommentTok{\# Plot the likelihoods}

\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow=}\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\FunctionTok{boxplot}\NormalTok{(}\AttributeTok{range =} \DecValTok{0}\NormalTok{,diff\_geneflow,recent\_geneflow,early\_geneflow,ongoing\_geneflow,}
\NormalTok{        no\_geneflow, }\AttributeTok{ylab=}\StringTok{"Likelihood"}\NormalTok{,}\AttributeTok{xaxt=}\StringTok{"n"}\NormalTok{)}
\FunctionTok{axis}\NormalTok{(}\AttributeTok{side=}\DecValTok{1}\NormalTok{,}\AttributeTok{at=}\DecValTok{1}\SpecialCharTok{:}\DecValTok{5}\NormalTok{, }\AttributeTok{labels=}\FunctionTok{c}\NormalTok{(}\StringTok{"early+recent"}\NormalTok{,}\StringTok{"recent"}\NormalTok{,}\StringTok{"early"}\NormalTok{,}\StringTok{"constant"}\NormalTok{,}\StringTok{"no"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

Similarly, we should compare the AIC values and see if AIC suggests that
the second-best model is significantly less good than the best model.
Given that we already used the calculateAIC.r script to compute AIC
values, this can easily be done.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{for i in */bestrun/*AIC}
\NormalTok{do}
\NormalTok{echo {-}e \textasciigrave{}basename $i\textasciigrave{}"\textbackslash{}t"\textasciigrave{}tail {-}n $i\textasciigrave{} \textgreater{}\textgreater{} allmodels.AIC}
\NormalTok{done}
\end{Highlighting}
\end{Shaded}

And we're done for this tutorial! It is worth noting that linkage among
SNPs can actually distort the AIC and likelihood calculations. As a
result it is best practice to perform block-bootstrapping to account for
this. We don't have time to implement that here but you can refer to the
\href{https://speciationgenomics.github.io/fastsimcoal2/}{original
version} of this tutorial which does have a guide for how to achieve
this.



\end{document}
