<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Land, River and Seascape Genomics - Future projections with GDM &amp; GFs</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Land, River and Seascape Genomics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./Teachers.html" rel="" target="">
 <span class="menu-text">Meet your teachers</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./Readings.html" rel="" target="">
 <span class="menu-text">Readings</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./resistance.html">Thursday</a></li><li class="breadcrumb-item"><a href="./GDM.html">Future projections with GDM &amp; GFs</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Sunday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Basic-Population-Genetics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic pop gen</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./basicR.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./basic-bash.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic Bash</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Monday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./land-river-sea.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Land, river, &amp; seascape genomics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./maps-spatial-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Maps &amp; spatial data</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Tuesday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genetic-dif.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Genetic differentiation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./RDA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Intro to RDA as a flexible tool</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Wednesday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sims_demo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slendr simulations</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Thursday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resistance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Resistance surfaces</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./GDM.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Future projections with GDM &amp; GFs</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./biophysical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Biophysical models</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">Friday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./GEA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">More GEA: genetic offsets &amp; maladaptation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./g_architectures.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Genomic architectures &amp; landscape genomics</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#geas-with-generalized-dissimilarity-modeling-gradient-forests" id="toc-geas-with-generalized-dissimilarity-modeling-gradient-forests" class="nav-link active" data-scroll-target="#geas-with-generalized-dissimilarity-modeling-gradient-forests">GEAs with Generalized dissimilarity modeling &amp; Gradient forests</a>
  <ul class="collapse">
  <li><a href="#gdm" id="toc-gdm" class="nav-link" data-scroll-target="#gdm">GDM</a>
  <ul class="collapse">
  <li><a href="#theorybackground-for-gdm" id="toc-theorybackground-for-gdm" class="nav-link" data-scroll-target="#theorybackground-for-gdm">Theory/Background for GDM</a></li>
  </ul></li>
  <li><a href="#gdm-applications-species-diversity" id="toc-gdm-applications-species-diversity" class="nav-link" data-scroll-target="#gdm-applications-species-diversity">GDM applications (species diversity)</a></li>
  <li><a href="#advantages-and-limitations-of-gdm" id="toc-advantages-and-limitations-of-gdm" class="nav-link" data-scroll-target="#advantages-and-limitations-of-gdm">Advantages and limitations of GDM</a>
  <ul class="collapse">
  <li><a href="#positives" id="toc-positives" class="nav-link" data-scroll-target="#positives">Positives</a></li>
  <li><a href="#drawbacks" id="toc-drawbacks" class="nav-link" data-scroll-target="#drawbacks">Drawbacks</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#gradient-forests" id="toc-gradient-forests" class="nav-link" data-scroll-target="#gradient-forests">Gradient forests</a>
  <ul class="collapse">
  <li><a href="#theory" id="toc-theory" class="nav-link" data-scroll-target="#theory">Theory</a>
  <ul class="collapse">
  <li><a href="#random-forests" id="toc-random-forests" class="nav-link" data-scroll-target="#random-forests">Random forests</a></li>
  <li><a href="#gradient-forests-1" id="toc-gradient-forests-1" class="nav-link" data-scroll-target="#gradient-forests-1">Gradient forests</a></li>
  </ul></li>
  <li><a href="#gf-applications-species-diversity" id="toc-gf-applications-species-diversity" class="nav-link" data-scroll-target="#gf-applications-species-diversity">GF applications (species diversity)</a></li>
  <li><a href="#advantages-and-limitations-of-gf" id="toc-advantages-and-limitations-of-gf" class="nav-link" data-scroll-target="#advantages-and-limitations-of-gf">Advantages and limitations of GF</a>
  <ul class="collapse">
  <li><a href="#positives-1" id="toc-positives-1" class="nav-link" data-scroll-target="#positives-1">Positives</a></li>
  <li><a href="#negatives" id="toc-negatives" class="nav-link" data-scroll-target="#negatives">Negatives</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#gdm-and-gf-applied-to-genomic-data" id="toc-gdm-and-gf-applied-to-genomic-data" class="nav-link" data-scroll-target="#gdm-and-gf-applied-to-genomic-data">GDM and GF applied to genomic data</a>
  <ul class="collapse">
  <li><a href="#the-data-set" id="toc-the-data-set" class="nav-link" data-scroll-target="#the-data-set">The data set</a></li>
  <li><a href="#results" id="toc-results" class="nav-link" data-scroll-target="#results">Results</a></li>
  <li><a href="#notes-developed-from" id="toc-notes-developed-from" class="nav-link" data-scroll-target="#notes-developed-from">Notes developed from:</a></li>
  </ul></li>
  <li><a href="#computer-tutorial" id="toc-computer-tutorial" class="nav-link" data-scroll-target="#computer-tutorial">Computer tutorial</a>
  <ul class="collapse">
  <li><a href="#overview" id="toc-overview" class="nav-link" data-scroll-target="#overview">Overview</a></li>
  <li><a href="#gdm-1" id="toc-gdm-1" class="nav-link" data-scroll-target="#gdm-1">GDM</a>
  <ul class="collapse">
  <li><a href="#fitting-and-examining-the-gdm-model" id="toc-fitting-and-examining-the-gdm-model" class="nav-link" data-scroll-target="#fitting-and-examining-the-gdm-model">Fitting and examining the GDM model</a></li>
  </ul></li>
  <li><a href="#gf" id="toc-gf" class="nav-link" data-scroll-target="#gf">GF</a>
  <ul class="collapse">
  <li><a href="#fitting-and-examining-the-gf-model" id="toc-fitting-and-examining-the-gf-model" class="nav-link" data-scroll-target="#fitting-and-examining-the-gf-model">Fitting and examining the GF model</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#points-for-class-discussion" id="toc-points-for-class-discussion" class="nav-link" data-scroll-target="#points-for-class-discussion">Points for class discussion</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Future projections with GDM &amp; GFs</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><em>Instructor: Riginos</em></p>
<section id="geas-with-generalized-dissimilarity-modeling-gradient-forests" class="level1">
<h1>GEAs with Generalized dissimilarity modeling &amp; Gradient forests</h1>
<ul>
<li><p>Developed for community ecology and adopted by landscape genetics (in community ecology, the biological data are species observations)</p></li>
<li><p>Focus on beta-diversity or turnover, show rates of biological change in similar (non linear) manners</p></li>
<li><p>Can predict genetic diversity across landscapes - environmental and geographic predictors can act as <em>surrogates</em> for turnover in genetic diversity (species distribution modeling makes the same assumption)</p></li>
<li><p>Being used to estimate genetic match/mismatch to future climate conditions</p></li>
<li><p>Uses distances: it is a link based analysis (second column in Wagner &amp; Fortin diagram)</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/statisticalModels_Wagner_Fortin_2016.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Statistical models - Wagner &amp; Fortin 2016</figcaption>
</figure>
</div>
<section id="gdm" class="level2">
<h2 class="anchored" data-anchor-id="gdm">GDM</h2>
<section id="theorybackground-for-gdm" class="level3">
<h3 class="anchored" data-anchor-id="theorybackground-for-gdm">Theory/Background for GDM</h3>
<section id="managing-non-linearity" class="level4">
<h4 class="anchored" data-anchor-id="managing-non-linearity">Managing non linearity</h4>
<p>GDM is <em>non</em> linear extension of matrix regression that uses dissimilarities in both predictor and response variables:</p>
<ul>
<li>Basic linear model: <span class="math inline">\(d_{ij} = a_0 + \sum_{p=1}^na_p|x_{pi}-x_{pj}|\)</span></li>
</ul>
<p>Modified to account for two aspects of non linearity:</p>
<ol type="1">
<li><p>Bounded compositional dissimilarity: <span class="math inline">\(d_{ij} \sim (0,1)\)</span></p></li>
<li><p>Variable rate of turnover</p></li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/nonlinearity_Ferrier_etal_2007.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Two types of non linearity - Ferrier et al 2007</figcaption>
</figure>
</div>
<p><strong>Solution for #1</strong>, use a generalized linear model that defines relationship between predictor (<span class="math inline">\(\eta\)</span>) and response (<span class="math inline">\(\mu\)</span>)</p>
<ul>
<li><p>Link function: <span class="math inline">\(\mu = 1 - e^{\eta}\)</span></p></li>
<li><p>Binomial variance: <span class="math inline">\(V(\mu) = \frac{\mu(\mu-1)}{s_i+s_j}\)</span></p></li>
</ul>
<p><strong>Solution for #2</strong>: fit nonlinear functions to the environmental variables, not their distances</p>
<ul>
<li><p>GLM equation is now changed to include functions for x</p>
<ul>
<li><span class="math inline">\(\eta = a_0 + \sum_{p=1}^n|f_p(x_{pi})-f_p(x_{pj})|\)</span></li>
</ul></li>
<li><p>an I-spline basis function is used as a function</p>
<ul>
<li><span class="math inline">\(f_p(x_p)= \sum_{k=1}^{m_p}a _{pk}I_{pk}(x_p)\)</span> (k splines per predictor)</li>
</ul></li>
<li><p>So, GLM becomes: <span class="math inline">\(\eta = a_0 + \sum_{p=1}^n\sum_{k=1}^{m_p}a_{pk}|I_{pk}(x_{pi})-I_{pk}(x_{pj})|\)</span></p></li>
</ul>
</section>
<section id="fitting-a-gdm-model" class="level4">
<h4 class="anchored" data-anchor-id="fitting-a-gdm-model">Fitting a GDM model</h4>
<ol type="1">
<li><p>compute <span class="math inline">\(d_{ij}\)</span> between all sites (n x n matrix)</p></li>
<li><p>derive a set of <span class="math inline">\(m_p\)</span> I-spline basis functions for each environmental variable and calculate value (<span class="math inline">\(I_{pk}(x_p)\)</span>)for each site</p></li>
<li><p>Calculate <span class="math inline">\(\Delta I_{pk}\)</span> for each pair of sites</p></li>
<li><p>For geographic and other distances, derive I-spline basis functions directly</p></li>
<li><p>Use maximum likelihood to fit coefficients</p></li>
<li><p>Contribution of each predictor evaluated by comparing <em>deviance</em> of model with and without the factor</p></li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/GDM_Mokany_etal_2022.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Assembling a GDM - Mokany et al 2022</figcaption>
</figure>
</div>
<p>Examining the I-splines</p>
<p><img src="images/I_splines_Mokany_etal_2022.png" class="img-fluid" alt="I-splines can show different rates of change - Mokany et al 2022">Uses of GDM</p>
</section>
</section>
</section>
<section id="gdm-applications-species-diversity" class="level2">
<h2 class="anchored" data-anchor-id="gdm-applications-species-diversity">GDM applications (species diversity)</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/GDM_usage_present_climate.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">GDM usage examples for present climate</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/GDMuseage_futureclimate.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">GDM usage examples for future climate</figcaption>
</figure>
</div>
</section>
<section id="advantages-and-limitations-of-gdm" class="level2">
<h2 class="anchored" data-anchor-id="advantages-and-limitations-of-gdm">Advantages and limitations of GDM</h2>
<section id="positives" class="level3">
<h3 class="anchored" data-anchor-id="positives">Positives</h3>
<ul>
<li><p>Very flexible on data types</p></li>
<li><p>Computation (for genetic analyses) does not scale with number of loci</p></li>
<li><p>Does not assume linearity</p></li>
<li><p>Can yield predictions for gridded landscapes</p></li>
</ul>
</section>
<section id="drawbacks" class="level3">
<h3 class="anchored" data-anchor-id="drawbacks">Drawbacks</h3>
<ul>
<li><p>Predictor distances are symmetric</p></li>
<li><p>Predictions in a large gridded landscape: computationally challenging (grid cell number choose 2 combinations)</p></li>
<li><p>Interactions between predictors are not modeled</p></li>
<li><p>Non independence of sites not accounted for if all sites used (Mokany et al.&nbsp;2022)</p></li>
<li><p>Alternative link functions might be more appropriate</p></li>
</ul>
</section>
</section>
</section>
<section id="gradient-forests" class="level1">
<h1>Gradient forests</h1>
<p><em>Gradient forests</em> is a machine learning approach based on <em>random forests</em>. Random forests are suited to predicting one outcome (such as the presences or absence of a specific species). Gradient forests puts together outcomes from many random forests (building information on a community of species or many loci).</p>
<section id="theory" class="level2">
<h2 class="anchored" data-anchor-id="theory">Theory</h2>
<section id="random-forests" class="level3">
<h3 class="anchored" data-anchor-id="random-forests">Random forests</h3>
<ul>
<li><p>Creates an ensemble of trees per species, these trees are split into partitions</p></li>
<li><p>Observations (sites for species, populations for genetics) are picked at random for each regression tree</p></li>
<li><p>For each node in the regression tree, predictor variables are selected at random and evaluated</p></li>
<li><p>Sites are split by a predictor value, s</p></li>
<li><p>Splits are chosen to create homogenous groups (measured by the sum of squared deviations about the group mean, called the <em>impurity</em>); the <em>importance</em> of each split is the reduction in impurity introduced by the split</p></li>
<li><p>Splitting occurs recursively until the tips are reached</p></li>
<li><p>Random forest = ensemble of trees where each trees is a bootstrap partition</p></li>
<li><p>Cross validation using “in bag” and “out of the bag” comparisons</p></li>
</ul>
<p>Confused? Me too! Let’s watch this video to get a better understanding…</p>
<iframe width="560" height="315" src="https://www.youtube.com/embed/J4Wdy0Wc_xQ?si=69YRNj86wmMjuCl6" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen="">
</iframe>
</section>
<section id="gradient-forests-1" class="level3">
<h3 class="anchored" data-anchor-id="gradient-forests-1">Gradient forests</h3>
<p>Combines ensembles of random forests to summarize importance of predictors and compositional turnover rates by predictor values.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/gradientforest_Ellis_etal_2012.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">How gradient forests work - Ellis et al 2012</figcaption>
</figure>
</div>
</section>
</section>
<section id="gf-applications-species-diversity" class="level2">
<h2 class="anchored" data-anchor-id="gf-applications-species-diversity">GF applications (species diversity)</h2>
<p>Example - outcomes for species abundance for species on the Great Barrier Reef</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/GF_example1_Ellis_etal_2012.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">GF outputs</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/GF_example2_Ellis_etal_2012.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">More GF outputs</figcaption>
</figure>
</div>
<p>Other applications</p>
<ul>
<li><p>Predicting biological turnover from environmental variables</p></li>
<li><p>Gap analyses of surveys</p></li>
<li><p>Identifying environmental thresholds to push community into a different state</p></li>
</ul>
</section>
<section id="advantages-and-limitations-of-gf" class="level2">
<h2 class="anchored" data-anchor-id="advantages-and-limitations-of-gf">Advantages and limitations of GF</h2>
<section id="positives-1" class="level3">
<h3 class="anchored" data-anchor-id="positives-1">Positives</h3>
<ul>
<li>Summarizes complex relationships with high predictive power - no variable transformation required</li>
<li>Species can be sampled in different ways and combined</li>
<li>Compositional dissimilarity is not bounded at 1</li>
<li>Correlated predictors should not affect results within the training region (but could if model is applied to new regions)</li>
</ul>
</section>
<section id="negatives" class="level3">
<h3 class="anchored" data-anchor-id="negatives">Negatives</h3>
<ul>
<li>Does not explicitly deal with spatial location (positions can be added as variables)</li>
<li>Small scale autocorrelation could cause problems - sample at distances larger than local SA</li>
</ul>
</section>
</section>
</section>
<section id="gdm-and-gf-applied-to-genomic-data" class="level1">
<h1>GDM and GF applied to genomic data</h1>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Fitzpatrick_Keller.png" class="img-fluid figure-img"></p>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Balsam-Poplar-300x225.jpg" class="img-fluid figure-img" alt="https://sierraclub.bc.ca/balsam-poplar/"></p>
<figcaption class="figure-caption">Balsam Poplar</figcaption>
</figure>
</div>
<p>Fitzpatrick &amp; Keller 2015 aimed to</p>
<ul>
<li><p>model how genetic diversity changes across a complex environmental landscape, especially for “adaptive” loci</p></li>
<li><p>project gene-environment relationships in space and time to yield landscape predictions and infer how environmental change will affect gene-environment associations</p></li>
</ul>
<p>This publication coined the term <strong><em>“genetic offset”</em></strong></p>
<blockquote class="blockquote">
<p>“The future projection of genetic composition informs us about how much the genetic composition across the landscape would have to change to preserve the gene–environment relationships observed under current environmental conditions.”</p>
</blockquote>
<blockquote class="blockquote">
<p>“Of course, the actual evolutionary responses of populations to environmental change will be more complex than this simpliﬁed projection, and likely to involve interactions between selection, the effective population size, which determines the efﬁcacy of selection response, and the evolutionary processes shaping adaptive variation (e.g.&nbsp;migration, mutation, recombination).”</p>
</blockquote>
<section id="the-data-set" class="level2">
<h2 class="anchored" data-anchor-id="the-data-set">The data set</h2>
<ul>
<li>Reference SNPs
<ul>
<li>412 SNPs = genomically random</li>
<li>31 populations</li>
<li>474 individuals</li>
</ul></li>
<li>Candidate loci
<ul>
<li>335 SNPS from flowering genes</li>
<li>443 individuals from 31 populations</li>
</ul></li>
<li>SNPs from specific candidate loci
<ul>
<li>previous outlier analyses identified GIGANTEA-5 (GI5), FRIGIDA (FRI) and LEAFY (LFY)</li>
</ul></li>
<li>Environmental variables: 6</li>
<li>Spatial structure (population history) accounted for by using geographic distance for GDM and MEMs for GF</li>
<li>GDM - response variables are pairwise Fst</li>
<li>GF - response variables are allele frequencies</li>
</ul>
</section>
<section id="results" class="level2">
<h2 class="anchored" data-anchor-id="results">Results</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/GDM_ISplines_Fitzpatrick_Keller_2015.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">GDM I-splines: Fitzpatrick &amp; Keller 2015</figcaption>
</figure>
</div>
<p>See Figure 2 in the original paper for similar GF results as above.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/GF_SNP_compositional_Turnover_Fitzpatrick_Keller_2015.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Turnover by SNP in response to summer temperature - Fitzpatrick &amp; Keller 2015</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Predicted _spatial_ Variation_Fitzpatrick_Keller_2015.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Predicted spatial variation - Fitzpatrick &amp; Keller 2015</figcaption>
</figure>
</div>
</section>
<section id="notes-developed-from" class="level2">
<h2 class="anchored" data-anchor-id="notes-developed-from">Notes developed from:</h2>
<blockquote class="blockquote">
<ul>
<li>Ellis, N., Smith, S. J., &amp; Pitcher, C. R. (2012). Gradient forests: calculating importance gradients on physical predictors. _Ecology, 93_(1), 156-168.</li>
<li>Ferrier, S., Manion, G., Elith, J., &amp; Richardson, K. (2007). Using generalized dissimilarity modelling to analyse and predict patterns of beta diversity in regional biodiversity assessment. _Diversity and Distributions, 13_(3), 252-264.</li>
<li>Fitzpatrick, M. C., &amp; Keller, S. R. (2015). Ecological genomics meets community-level modelling of biodiversity: mapping the genomic landscape of current and future environmental adaptation. _Ecol Lett, 18_(1), 1-16.</li>
<li>Mokany, K., Ware, C., Woolley, S. N. C., Ferrier, S., Fitzpatrick, Matthew&nbsp;C., &amp; Bahn, V. (2022). A working guide to harnessing generalized dissimilarity modelling for biodiversity analysis and conservation assessment. _Global Ecology and Biogeography, 31_(4), 802-821.</li>
</ul>
</blockquote>
<hr>
</section>
</section>
<section id="computer-tutorial" class="level1">
<h1>Computer tutorial</h1>
<section id="overview" class="level2">
<h2 class="anchored" data-anchor-id="overview">Overview</h2>
<p>This tutorial will provide the briefest introduction to GDMs and GF’s and their usage in landscape genomics.</p>
<p>These methods are included because the represent different ways of approaching landscape data and are often encountered in genomic prediction studies.</p>
<p>The scripts support the analyses presented in Fitzpatrick and Keller 2015 on balsam poplar and are based on the scripts and data found on <a href="">Data Dryad</a>.</p>
<p>Unfortunately the authors did not make their spatial data layers available (likely due to large file sizes) and therefore we cannot recapitulate their landscape projections to look at genomic offsets. We will do this later based on an RDA method.</p>
<p>As we have encountered before, support for the packages <code>GDM</code> and <code>gradientForest</code> might not be fully up to date with developments in R spatial data handling.</p>
<p>All data from: Fitzpatrick, M. C., &amp; Keller, S. R. (2015). Ecological genomics meets community-level modelling of biodiversity: mapping the genomic landscape of current and future environmental adaptation. Ecol Lett, 18(1), 1-16.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("gradientForest", repos="http://R-Forge.R-project.org")</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gdm) <span class="co">#notice the warnings, we can talk about this</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gradientForest)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(raster)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="gdm-1" class="level2">
<h2 class="anchored" data-anchor-id="gdm-1">GDM</h2>
<p>Remember that GDM is based on distances between sites. If you examine the <code>gdmData</code> file you will see that each row contains information regarding the relationship between a pair of sites.</p>
<ul>
<li><p>Genetic data - Representing the genetic response variables are FST values for various sets of data (reference loci, candidate, GI5, etc.) These Fst values have been scaled and centered.</p></li>
<li><p>Coordinates - The locations of each population pair are included as x and y coordinates.</p></li>
<li><p>Environmental distances - expressed as <span class="math inline">\(\Delta I_{pk}\)</span> values - these have been calculated already by running the function <code>formatsitepair()</code>. If you were preparing your own data, you would need to run this function to fit the splines.</p></li>
</ul>
<section id="fitting-and-examining-the-gdm-model" class="level3">
<h3 class="anchored" data-anchor-id="fitting-and-examining-the-gdm-model">Fitting and examining the GDM model</h3>
<p>We will focus on the reference and GIGANTEA-5 (GI5) datasets. To fit the models, individual dataframes need to be constructed.</p>
<p>The datafile should be preloaded, otherwise download from <a href="data/poplarFst.ENV.data.4.GDM.csv">here</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># import GDM ready table</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>gdmData <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"./data/poplarFst.ENV.data.4.GDM.csv"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># build individual SNP datasets</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>SNPs_ref <span class="ot">&lt;-</span> gdmData[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">6</span><span class="sc">:</span><span class="dv">24</span>)] <span class="co"># reference</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>GI5 <span class="ot">&lt;-</span> gdmData[,<span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">6</span><span class="sc">:</span><span class="dv">24</span>)] <span class="co"># GIGANTEA-5 (GI5)</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="co"># fit and plot GDM for reference SNPs</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>gdmRef <span class="ot">&lt;-</span> <span class="fu">gdm</span>(SNPs_ref, <span class="at">geo=</span><span class="cn">TRUE</span>) <span class="co">#gives warning but works</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>gdmRef<span class="sc">$</span>explained <span class="co"># From F&amp;K2015: "GDM explained more than 63% of the deviance in turnover in genetic composition of reference"</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(gdmRef)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>refSplines <span class="ot">&lt;-</span> <span class="fu">isplineExtract</span>(gdmRef) <span class="co"># extract spline data for custom plotting</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="co"># fit and plot GDM for GI5 SNPs</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>gdmGI5 <span class="ot">&lt;-</span> <span class="fu">gdm</span>(GI5, <span class="at">geo=</span><span class="cn">TRUE</span>) <span class="co">#ignore warning</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>gdmGI5<span class="sc">$</span>explained</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(gdmGI5)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You will recognize some of the plots from Figure 3. Notice how variable the spline functions can be.</p>
</section>
</section>
<section id="gf" class="level2">
<h2 class="anchored" data-anchor-id="gf">GF</h2>
<p>In contrast to GDM, GF is site based. Therefore each row in the dataframe corresponds to a site (population).</p>
<ul>
<li><p>Coordinates - The locations of each population are included as x and y coordinates.</p></li>
<li><p>Environmental values - Appear to be raw values.</p></li>
<li><p>MEMS - are included as additional spatial variables.</p></li>
<li><p>Genetic data - Are allele frequencies. (Columns 14-373 are reference loci =360 loci, does not match the paper!)</p></li>
</ul>
<section id="fitting-and-examining-the-gf-model" class="level3">
<h3 class="anchored" data-anchor-id="fitting-and-examining-the-gf-model">Fitting and examining the GF model</h3>
<p>As before, we focus on the reference and GIGANTEA-5 (GI5) datasets. And make individual dataframes to feed into the analyses.</p>
<p>The datafile should be preloaded, otherwise download from <a href="data/poplarSNP.ENV.data.4.GF.csv">here</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>gfData <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"./data/poplarSNP.ENV.data.4.GF.csv"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>envGF <span class="ot">&lt;-</span> gfData[,<span class="dv">3</span><span class="sc">:</span><span class="dv">13</span>] <span class="co"># get climate &amp; MEM variables</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># build individual SNP datasets</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>SNPs_ref <span class="ot">&lt;-</span> gfData[,<span class="fu">grep</span>(<span class="st">"REFERENCE"</span>,<span class="fu">colnames</span>(gfData))] <span class="co"># reference 360 loci</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>GI5 <span class="ot">&lt;-</span> gfData[,<span class="fu">grep</span>(<span class="st">"GI5"</span>,<span class="fu">colnames</span>(gfData))] <span class="co"># GIGANTEA-5 (GI5)</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>maxLevel <span class="ot">&lt;-</span> <span class="fu">log2</span>(<span class="fl">0.368</span><span class="sc">*</span><span class="fu">nrow</span>(envGF)<span class="sc">/</span><span class="dv">2</span>) <span class="co">#account for correlations, F&amp;K set this flag but do not explain beyond saying "see ?gradientForest"</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit gf models for reference SNPs </span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>gfRef <span class="ot">&lt;-</span> <span class="fu">gradientForest</span>(<span class="fu">cbind</span>(envGF, SNPs_ref), <span class="at">predictor.vars=</span><span class="fu">colnames</span>(envGF), <span class="at">response.vars=</span><span class="fu">colnames</span>(SNPs_ref), <span class="at">ntree=</span><span class="dv">500</span>, <span class="at">maxLevel=</span>maxLevel, <span class="at">trace=</span>T, <span class="at">corr.threshold=</span><span class="fl">0.50</span>)</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit gf models for GI5 SNPs</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>gfGI5 <span class="ot">&lt;-</span> <span class="fu">gradientForest</span>(<span class="fu">cbind</span>(envGF, GI5), <span class="at">predictor.vars=</span><span class="fu">colnames</span>(envGF),</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>                          <span class="at">response.vars=</span><span class="fu">colnames</span>(GI5), <span class="at">ntree=</span><span class="dv">500</span>, </span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>                          <span class="at">maxLevel=</span>maxLevel, <span class="at">trace=</span>T, <span class="at">corr.threshold=</span><span class="fl">0.50</span>)</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="co"># plot output, see ?plot.gradientForest</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(gfRef, <span class="at">plot.type=</span><span class="st">"O"</span>) <span class="co"># values that feed into Fig1</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(gfRef, <span class="at">plot.type =</span> <span class="st">"C"</span>)  <span class="co"># Fig2 </span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(gfRef, <span class="at">plot.type =</span> <span class="st">"S"</span>)</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>     </span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="co">#can do the same for GI5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
</section>
<section id="points-for-class-discussion" class="level1">
<h1>Points for class discussion</h1>
<ol type="1">
<li><p>What are the points of difference and similarity between GDM and GF? Which might you use in what context?</p></li>
<li><p>The GDM example used pairwise FST values. Does this seem like a reasonable choice? What implicit assumptions go into this choice?</p></li>
<li><p>There were major reproducibility issues in trying to reconstruct the analyses! Spatial data sets are missing, packages are old and not necessarily updated…. what best practices could we follow to mitigate such problems?</p></li>
</ol>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>