<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Land, River and Seascape Genomics - More GEA: genetic offsets &amp; maladaptation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Land, River and Seascape Genomics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./Teachers.html" rel="" target="">
 <span class="menu-text">Meet your teachers</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./Readings.html" rel="" target="">
 <span class="menu-text">Readings</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./GEA.html">Friday</a></li><li class="breadcrumb-item"><a href="./GEA.html">More GEA: genetic offsets &amp; maladaptation</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Sunday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Basic-Population-Genetics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic pop gen</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./basicR.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./basic-bash.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic Bash</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Monday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./land-river-sea.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Land, river, &amp; seascape genomics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./maps-spatial-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Maps &amp; spatial data</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Tuesday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genetic-dif.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Genetic differentiation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./RDA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Intro to RDA as a flexible tool</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Wednesday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sims_demo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slendr simulations</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Thursday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resistance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Resistance surfaces</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./GDM.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Future projections with GDM &amp; GFs</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./biophysical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Biophysical models</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">Friday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./GEA.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">More GEA: genetic offsets &amp; maladaptation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./g_architectures.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Genomic architectures &amp; landscape genomics</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#under-construction" id="toc-under-construction" class="nav-link active" data-scroll-target="#under-construction">Under construction!!</a>
  <ul class="collapse">
  <li><a href="#the-general-genomic-offset-framework-is-as-follows" id="toc-the-general-genomic-offset-framework-is-as-follows" class="nav-link" data-scroll-target="#the-general-genomic-offset-framework-is-as-follows">The general genomic offset framework is as follows:</a>
  <ul class="collapse">
  <li><a href="#more-details---genomic-prediction" id="toc-more-details---genomic-prediction" class="nav-link" data-scroll-target="#more-details---genomic-prediction">More details - <em>genomic prediction</em></a></li>
  <li><a href="#more-details---genomic-offset" id="toc-more-details---genomic-offset" class="nav-link" data-scroll-target="#more-details---genomic-offset">More details - <em>genomic offset</em></a></li>
  <li><a href="#adding-in-dispersal" id="toc-adding-in-dispersal" class="nav-link" data-scroll-target="#adding-in-dispersal">Adding in dispersal …</a></li>
  <li><a href="#validation" id="toc-validation" class="nav-link" data-scroll-target="#validation">Validation</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#references-for-this-tutorial" id="toc-references-for-this-tutorial" class="nav-link" data-scroll-target="#references-for-this-tutorial">References for this tutorial</a></li>
  <li><a href="#computer-tutorial---geas-from-rdas" id="toc-computer-tutorial---geas-from-rdas" class="nav-link" data-scroll-target="#computer-tutorial---geas-from-rdas">Computer tutorial - GEAs from RDAs</a>
  <ul class="collapse">
  <li><a href="#identifying-loci-under-selection" id="toc-identifying-loci-under-selection" class="nav-link" data-scroll-target="#identifying-loci-under-selection">Identifying loci under selection</a>
  <ul class="collapse">
  <li><a href="#load-libraries-and-data-files" id="toc-load-libraries-and-data-files" class="nav-link" data-scroll-target="#load-libraries-and-data-files">Load libraries and data files</a></li>
  <li><a href="#partial-rda-for-identified-environmental-variables" id="toc-partial-rda-for-identified-environmental-variables" class="nav-link" data-scroll-target="#partial-rda-for-identified-environmental-variables">Partial RDA for identified environmental variables</a></li>
  <li><a href="#identifying-outliers-based-on-mahalanobis-distances-from-center-of-rda-space" id="toc-identifying-outliers-based-on-mahalanobis-distances-from-center-of-rda-space" class="nav-link" data-scroll-target="#identifying-outliers-based-on-mahalanobis-distances-from-center-of-rda-space">Identifying outliers based on Mahalanobis distances from center of RDA space</a></li>
  </ul></li>
  <li><a href="#recreating-fig-1a" id="toc-recreating-fig-1a" class="nav-link" data-scroll-target="#recreating-fig-1a">Recreating Fig 1a</a></li>
  <li><a href="#recreating-fig-1b" id="toc-recreating-fig-1b" class="nav-link" data-scroll-target="#recreating-fig-1b">Recreating Fig 1b</a></li>
  <li><a href="#simple-rda-for-outlier-identification" id="toc-simple-rda-for-outlier-identification" class="nav-link" data-scroll-target="#simple-rda-for-outlier-identification">Simple RDA for outlier identification</a>
  <ul class="collapse">
  <li><a href="#overlap-among-all-outliers-with-and-without-conditioning-on-population-structure" id="toc-overlap-among-all-outliers-with-and-without-conditioning-on-population-structure" class="nav-link" data-scroll-target="#overlap-among-all-outliers-with-and-without-conditioning-on-population-structure">Overlap among all outliers with and without conditioning on population structure</a></li>
  <li><a href="#overlap-among-rda-top-outliers-with-and-without-conditioning-on-population-structure" id="toc-overlap-among-rda-top-outliers-with-and-without-conditioning-on-population-structure" class="nav-link" data-scroll-target="#overlap-among-rda-top-outliers-with-and-without-conditioning-on-population-structure">Overlap among RDA (“top”) outliers with and without conditioning on population structure</a></li>
  </ul></li>
  <li><a href="#spatial-projection-to-make-map-figures" id="toc-spatial-projection-to-make-map-figures" class="nav-link" data-scroll-target="#spatial-projection-to-make-map-figures">Spatial projection to make map figures</a></li>
  </ul></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="GEA.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">More GEA: genetic offsets &amp; maladaptation</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><em>Instructor: Riginos</em></p>
<section id="under-construction" class="level1">
<h1>Under construction!!</h1>
<p>One of the goals of landscape genomics is to find loci that contribute to adaptation.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/environment_genotype_phenotype_Rellstab2015.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Rellstab et al 2015 (Image modified from Sork et al 2013)</figcaption>
</figure>
</div>
<p>Tests of selection, especially based on finding loci that show strong correlations with environmental features, can provide lists of candidate loci.</p>
<p>Simulations indicate that RDA-based methods of identifying outliers perform well as tests of environmentally induced selection (Forester et al 2016).</p>
<p>Given the unfolding climate crisis, there is growing interest in leveraging genetic information to identify populations that are vulnerable (or resilient) to anticipated future climate conditions.</p>
<p>There are numerous names for this concept:</p>
<p>(mal)adaptation = risk of non-adaptedness = genomic offset = genomic vulnerability</p>
<section id="the-general-genomic-offset-framework-is-as-follows" class="level2">
<h2 class="anchored" data-anchor-id="the-general-genomic-offset-framework-is-as-follows">The general genomic offset framework is as follows:</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/GEAframework_Capblancq_etal_2020.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">GEA framework - Capblancq et al.&nbsp;2020</figcaption>
</figure>
</div>
<p>Capblancq et al.&nbsp;2020</p>
<p>For example, in the lodgepole pine data used by Capblancq &amp; Forester (2021), they:</p>
<ol type="1">
<li><p>used RDAs to identify correlations between environmental variables and allele frequencies</p></li>
<li><p>identified outlier loci showing the strongest associations with environmental variables</p></li>
<li><p>projected locally “adaptive” alleles across the landscape: this is the predicted genetic score per grid cell based on the environmental scores, or <em>genomic prediction</em></p></li>
<li><p>projected the future adaptive value “needed” at each grid cell based on future environment and then calculated the difference between present and future as the <em>genomic offset</em></p></li>
<li><p>(look to independent validation)</p></li>
</ol>
<section id="more-details---genomic-prediction" class="level3">
<h3 class="anchored" data-anchor-id="more-details---genomic-prediction">More details - <em>genomic prediction</em></h3>
<p>Basic framework:</p>
<p>environmental data -&gt; run through GEA model()-&gt; predicted genetic values</p>
<p>Fitzpatrick and Keller 2015 example using gradient forests on balsam poplar:</p>
<ul>
<li><p>Panels A &amp; B are genomic predictions</p></li>
<li><p>Panel C is “distance” of G15 from neutral population structure</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/genomic_prediction_Fitzpatrick_Keller_2015.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Genomic prediction - Fitzpatrick &amp; Keller 2015</figcaption>
</figure>
</div>
<p>In the C&amp;F example using RDA on lodgepole pine:</p>
<p>“Adaptive index” is calculated per grid cell as:</p>
<p>Adaptive index per grid cell = <span class="math inline">\(\sum\limits_{i=1}^{n}a_ib_i\)</span>, where</p>
<ul>
<li><span class="math inline">\(a_i\)</span> is the loading on the RDA axis (the GEA model)</li>
<li><span class="math inline">\(b_i\)</span> is the environmental variable at that grid cell</li>
<li><span class="math inline">\(n\)</span> is the number of environmental variables</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/mapping_RDAs_Capblancq_Forester_2021.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Adaptive index - Capblancq &amp; Forester 2022</figcaption>
</figure>
</div>
</section>
<section id="more-details---genomic-offset" class="level3">
<h3 class="anchored" data-anchor-id="more-details---genomic-offset">More details - <em>genomic offset</em></h3>
<p>Basic framework:</p>
<p>GO = predicted genetic values at present - predicted genetic values for future</p>
<p>Fitzpatrick and Keller 2015 example using gradient forests on balsam poplar:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/genomic_offset_Fitzpatrick_Keller_2015.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Genomic offset - Fitzpatrick &amp; Keller 2015</figcaption>
</figure>
</div>
<p>In the C&amp;F example using RDA on lodgepole pine:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/genomicoffsets_Capblancq_Forester_2021.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Genomic offset - Capblancq &amp; Forester 2022</figcaption>
</figure>
</div>
</section>
<section id="adding-in-dispersal" class="level3">
<h3 class="anchored" data-anchor-id="adding-in-dispersal">Adding in dispersal …</h3>
<p>(based on Gougherty et al.&nbsp;2021)</p>
<p>At the landscape level, how different will environments be and how geographically distant?</p>
<ul>
<li><p>Forward offsets - what is minimal Fst distance from present day to future? Where would present populations need to move to?</p></li>
<li><p>Reverse offsets - what is minimal Fst distance from future to present? Where would future populations need to come from?</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Forward_reverse_offsets_Gougherty_etal_2021.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Concepts of forward and reverse offsets - Gougherty etal 2021</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Forward_reverse_offsets2_Gougherty_etal_2021.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Migration distance to acheive offsets - Gougherty etal 2021</figcaption>
</figure>
</div>
<ul>
<li>Assisted migration would be needed for eastern populations</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Forward_offset_distances_Gougherty_etal_2021.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Migration distances - Gougherty etal 2021</figcaption>
</figure>
</div>
<ul>
<li>But most forward offset migration could be achieved in 500 km</li>
</ul>
</section>
<section id="validation" class="level3">
<h3 class="anchored" data-anchor-id="validation">Validation</h3>
<p>Are we confident that GEA analyses:</p>
<ol type="1">
<li>Can identify important environmental variables for shaping genetic diversity ?</li>
</ol>
<ul>
<li>Cautiously, yes</li>
</ul>
<ol start="2" type="1">
<li>Identify genetic loci that contributed to relevant phenotypes?</li>
</ol>
<ul>
<li>Limited empirical studies that compare results across methods (plants, mostly) - some consistent results for individual loci (that likely have large effects)</li>
</ul>
<p>See Lasky et al 2023 for a good review.</p>
</section>
</section>
</section>
<section id="references-for-this-tutorial" class="level1">
<h1>References for this tutorial</h1>
<ul>
<li><p>Capblancq, T., Fitzpatrick, M. C., Bay, R. A., Exposito-Alonso, M., &amp; Keller, S. R. (2020). Genomic prediction of (mal)adaptation across current and future climatic landscapes. Annual Review of Ecology, Evolution, and Systematics, 51(1), 245-269. doi:10.1146/annurev-ecolsys-020720-042553</p></li>
<li><p>Capblancq, T., &amp; Forester, B. R. (2021). Redundancy analysis: A Swiss Army Knife for landscape genomics. Methods in Ecology and Evolution. doi:10.1111/2041-210x.13722</p></li>
<li><p>Forester, B. R., Jones, M. R., Joost, S., Landguth, E. L., &amp; Lasky, J. R. (2016). Detecting spatial genetic signatures of local adaptation in heterogeneous landscapes. Molecular Ecology, 25(1), 104-120. doi: 10.1111/mec.13476</p></li>
<li><p>Gougherty, A. V., Keller, S. R., &amp; Fitzpatrick, M. C. (2021). Maladaptation, migration and extirpation fuel climate change risk in a forest tree species. _Nat Clim Change, 11_(2), 166-171.</p></li>
<li><p>Lasky, J. R., Josephs, E. B., &amp; Morris, G. P. (2023). Genotype-environment associations to reveal the molecular basis of environmental adaptation. Plant Cell, 35, 125-138. doi:10.1093/plcell/koac267</p></li>
</ul>
</section>
<section id="computer-tutorial---geas-from-rdas" class="level1">
<h1>Computer tutorial - GEAs from RDAs</h1>
<p>There are many ways that GEA analyses can be done. As a demonstration of one approach, we will continue with the tutorial from Capblancq &amp; Forrester 2021 using lodgepole pine data. As with the RDA tutorial, original comments from C&amp;F will be</p>
<blockquote class="blockquote">
<p>indented like this.</p>
</blockquote>
<p>If you remember how RDA works, one of the outputs of the RDA is the loadings (or “scores”) per locus. The goal is to identify the loci that show the strongest association or loadings with environmental attributes. In this example, the loci with the highest loadings to RDA axes (representing a composite of predictors) will be identified.</p>
<section id="identifying-loci-under-selection" class="level2">
<h2 class="anchored" data-anchor-id="identifying-loci-under-selection">Identifying loci under selection</h2>
<p>The first task is to identify loci showing strong associations with environmental features that are <em>candidates</em> for selection. We start by reloading some of the files we worked with in the previous tutorial.</p>
<section id="load-libraries-and-data-files" class="level3">
<h3 class="anchored" data-anchor-id="load-libraries-and-data-files">Load libraries and data files</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggVennDiagram)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># required for pdadapt.R function</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(robust)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co"># To install qvalue on your own device: </span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="co">#if (!require("BiocManager", quietly = TRUE))</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="co">#    install.packages("BiocManager")</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co">#BiocManager::install("qvalue")</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(qvalue)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>path<span class="ot">&lt;-</span><span class="st">"/home/data/gea/"</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Genetic data</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="fu">paste0</span>(path, <span class="st">"AllFreq.RData"</span>))</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="fu">paste0</span>(path, <span class="st">"Neutral.RData"</span>))</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Spatial data - files that you made in RDA workshop</span></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>Variables<span class="ot">&lt;-</span><span class="fu">read.csv</span>(<span class="fu">paste0</span>(path, <span class="st">"Variables.csv"</span>))</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>Coordinates<span class="ot">&lt;-</span><span class="fu">read.csv</span>(<span class="fu">paste0</span>(path, <span class="st">"Coordinates.csv"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>The first step was to run an RDA model on the allele frequency matrix using the 9 retained climatic factors as explanatory variables and the first three PCs as conditioning variables to account for neutral population structure.</p>
</blockquote>
</section>
<section id="partial-rda-for-identified-environmental-variables" class="level3">
<h3 class="anchored" data-anchor-id="partial-rda-for-identified-environmental-variables">Partial RDA for identified environmental variables</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>RDA_env <span class="ot">&lt;-</span> <span class="fu">rda</span>(AllFreq <span class="sc">~</span> MAR <span class="sc">+</span> EMT <span class="sc">+</span> MWMT <span class="sc">+</span> CMD <span class="sc">+</span> Tave_wt <span class="sc">+</span> DD_18 <span class="sc">+</span> MAP <span class="sc">+</span> Eref <span class="sc">+</span> PAS <span class="sc">+</span> <span class="fu">Condition</span>(PC1 <span class="sc">+</span> PC2 <span class="sc">+</span> PC3),  Variables)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We then had to choose a number of RDA axes to include when conducting the genome scan.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">screeplot</span>(RDA_env, <span class="at">main=</span><span class="st">"Eigenvalues of constrained axes"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>Looking at the proportion of variance explained by each axis, we decided to use the first two axes in this demonstration, though retaining just the first axis, or axes 1-3 would be other reasonable choices. If selecting only one axis, the <a href="https://popgen.nescent.org/2018-03-27_RDA_GEA.html">axis-based approach</a> mentioned above would be required to identify outliers.</p>
</blockquote>
<blockquote class="blockquote">
<p><strong>Note:</strong> It can be interesting to calculate the proportion of variance explained by the RDA model and the individual RDA axes when conducting a GEA test (code available <a href="https://popgen.nescent.org/2018-03-27_RDA_GEA.html">here</a>); however having a high proportion of variance explained is not required for GEA testing, which is focused on identifying outlier loci in the ordination space, not maximizing variance explained.</p>
</blockquote>
</section>
<section id="identifying-outliers-based-on-mahalanobis-distances-from-center-of-rda-space" class="level3">
<h3 class="anchored" data-anchor-id="identifying-outliers-based-on-mahalanobis-distances-from-center-of-rda-space">Identifying outliers based on Mahalanobis distances from center of RDA space</h3>
<p>Outlier loci will be the most extreme from the center of the distribution - see C&amp;F Fig1a that we will be reproducing here.</p>
<blockquote class="blockquote">
<p>For the lodgepole pine data, we will load the <a href="XXX">rdadapt</a> function, described in Capblancq et al.&nbsp;(2018) and use it to conduct the genome scan.</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Function rdadapt</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">paste0</span>(path, <span class="st">"rdadapt.R"</span>))</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Running the function with K = 2</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>rdadapt_env<span class="ot">&lt;-</span><span class="fu">rdadapt</span>(RDA_env, <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>One critical step when conducting a genome scan is to set a pertinent p-value threshold to identify the outlier loci. Here, we used a Bonferroni correction to account for multiple testing.</p>
</blockquote>
<blockquote class="blockquote">
<p><strong>Note:</strong> the rdadapt function returns both p-values and q-values, which means it is possible to use a FDR (False Discovery Rate) approach instead of a p-value threshold to identify outliers; see <a href="https://onlinelibrary.wiley.com/doi/full/10.1111/mec.13513">Francois et al.&nbsp;2015</a>.</p>
</blockquote>
<p>Some strategies for trying to understand other people’s code… this custom function had <strong>no (!)</strong> comments in it. To understand other people’s functions, I try to work through them. Open the file and take a look (all comments are from CR). To understand this function, I define inputs based on our examples, namely, <code>rda&lt;-RDA_env</code> and <code>K&lt;-2</code> and then execute lines 4-11 in the function code piece by piece and try to understand what is going on. (And add my own comments in case I need to go back to the code… )</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">## P-values threshold after Bonferroni correction</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>thres_env <span class="ot">&lt;-</span> <span class="fl">0.01</span><span class="sc">/</span><span class="fu">length</span>(rdadapt_env<span class="sc">$</span>p.values)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Identifying the loci that are below the p-value threshold</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>outliers <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">Loci =</span> <span class="fu">colnames</span>(AllFreq)[<span class="fu">which</span>(rdadapt_env<span class="sc">$</span>p.values<span class="sc">&lt;</span>thres_env)], </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">p.value =</span> rdadapt_env<span class="sc">$</span>p.values[<span class="fu">which</span>(rdadapt_env<span class="sc">$</span>p.values<span class="sc">&lt;</span>thres_env)], </span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">contig =</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(<span class="fu">colnames</span>(AllFreq)[<span class="fu">which</span>(rdadapt_env<span class="sc">$</span>p.values<span class="sc">&lt;</span>thres_env)], <span class="at">split =</span> <span class="st">"_"</span>), </span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) x[<span class="dv">1</span>]))</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="co"># that is one ungly piece of code...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>(Creation of <code>outliers</code> code above highlights how subsetting in base R can be difficult to read as a human… tidyverse syntax would be helpful…)</p>
<blockquote class="blockquote">
<p>To avoid redundancy among loci that are in strong physical linkage along the genome we only kept the outlier locus with the lowest p-values for each genomic contig.</p>
</blockquote>
<blockquote class="blockquote">
<p><strong>Note:</strong> To avoid linkage disequilibrium among loci, it is also possible to prune the dataset before running the genome scan. Here we preferred not to to be able to compare our top hit outliers with previously found adaptive loci that were searched for on a non-pruned dataset.</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Top hit outlier per contig</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>outliers <span class="ot">&lt;-</span> outliers[<span class="fu">order</span>(outliers<span class="sc">$</span>contig, outliers<span class="sc">$</span>p.value),]</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="do">## List of outlier names</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>outliers_rdadapt_env <span class="ot">&lt;-</span> <span class="fu">as.character</span>(outliers<span class="sc">$</span>Loci[<span class="sc">!</span><span class="fu">duplicated</span>(outliers<span class="sc">$</span>contig)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="recreating-fig-1a" class="level2">
<h2 class="anchored" data-anchor-id="recreating-fig-1a">Recreating Fig 1a</h2>
<p>Once the outliers have been identified, it can be useful to visualize their distribution in comparison with neutral loci using either an RDA biplot or a Manhattan plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Formatting table for ggplot</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>locus_scores <span class="ot">&lt;-</span> <span class="fu">scores</span>(RDA_env, <span class="at">choices=</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>), <span class="at">display=</span><span class="st">"species"</span>, <span class="at">scaling=</span><span class="st">"none"</span>) <span class="co"># vegan references "species", here these are the loci</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Create category for type of loci</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Top outiers are the outliers from the pRDA</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>TAB_loci <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">names =</span> <span class="fu">row.names</span>(locus_scores), locus_scores)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>TAB_loci<span class="sc">$</span>type <span class="ot">&lt;-</span> <span class="st">"Neutral"</span> <span class="co">#make all as neutral </span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>TAB_loci<span class="sc">$</span>type[TAB_loci<span class="sc">$</span>names<span class="sc">%in%</span>outliers<span class="sc">$</span>Loci] <span class="ot">&lt;-</span> <span class="st">"All outliers"</span> <span class="co">#overwrite select types</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>TAB_loci<span class="sc">$</span>type[TAB_loci<span class="sc">$</span>names<span class="sc">%in%</span>outliers_rdadapt_env] <span class="ot">&lt;-</span> <span class="st">"Top outliers"</span> <span class="co">#overwrite select types</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>TAB_loci<span class="sc">$</span>type <span class="ot">&lt;-</span> <span class="fu">factor</span>(TAB_loci<span class="sc">$</span>type, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Neutral"</span>, <span class="st">"All outliers"</span>, <span class="st">"Top outliers"</span>)) <span class="co">#make as factor</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>TAB_loci <span class="ot">&lt;-</span> TAB_loci[<span class="fu">order</span>(TAB_loci<span class="sc">$</span>type),] <span class="co">#reorder</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>TAB_var <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">scores</span>(RDA_env, <span class="at">choices=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">display=</span><span class="st">"bp"</span>)) <span class="co"># pull the biplot scores</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="do">## Biplot of RDA loci and variables scores</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept=</span><span class="dv">0</span>, <span class="at">linetype=</span><span class="st">"dashed"</span>, <span class="at">color =</span> <span class="fu">gray</span>(.<span class="dv">80</span>), <span class="at">linewidth=</span><span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept=</span><span class="dv">0</span>, <span class="at">linetype=</span><span class="st">"dashed"</span>, <span class="at">color =</span> <span class="fu">gray</span>(.<span class="dv">80</span>), <span class="at">linewidth=</span><span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> TAB_loci, <span class="fu">aes</span>(<span class="at">x=</span>RDA1<span class="sc">*</span><span class="dv">20</span>, <span class="at">y=</span>RDA2<span class="sc">*</span><span class="dv">20</span>, <span class="at">colour =</span> type), <span class="at">size =</span> <span class="fl">1.4</span>) <span class="sc">+</span></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"gray90"</span>, <span class="st">"#F9A242FF"</span>, <span class="st">"#6B4596FF"</span>)) <span class="sc">+</span></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="at">data =</span> TAB_var, <span class="fu">aes</span>(<span class="at">xend=</span>RDA1, <span class="at">yend=</span>RDA2, <span class="at">x=</span><span class="dv">0</span>, <span class="at">y=</span><span class="dv">0</span>), <span class="at">colour=</span><span class="st">"black"</span>, <span class="at">size=</span><span class="fl">0.15</span>, <span class="at">linetype=</span><span class="dv">1</span>, <span class="at">arrow=</span><span class="fu">arrow</span>(<span class="at">length =</span> <span class="fu">unit</span>(<span class="fl">0.02</span>, <span class="st">"npc"</span>))) <span class="sc">+</span></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> TAB_var, <span class="fu">aes</span>(<span class="at">x=</span><span class="fl">1.1</span><span class="sc">*</span>RDA1, <span class="at">y=</span><span class="fl">1.1</span><span class="sc">*</span>RDA2, <span class="at">label =</span> <span class="fu">row.names</span>(TAB_var)), <span class="at">size =</span> <span class="fl">2.5</span>, <span class="at">family =</span> <span class="st">"Times"</span>) <span class="sc">+</span></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"RDA 1"</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"RDA 2"</span>) <span class="sc">+</span></span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span><span class="st">"RDA space"</span>) <span class="sc">+</span></span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color=</span><span class="fu">guide_legend</span>(<span class="at">title=</span><span class="st">"Locus type"</span>)) <span class="sc">+</span></span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">11</span>, <span class="at">base_family =</span> <span class="st">"Times"</span>) <span class="sc">+</span></span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.background =</span> <span class="fu">element_blank</span>(), <span class="at">legend.background =</span> <span class="fu">element_blank</span>(), <span class="at">panel.grid =</span> <span class="fu">element_blank</span>(), <span class="at">plot.background =</span> <span class="fu">element_blank</span>(), <span class="at">legend.text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(.<span class="dv">8</span>)), <span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">11</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="recreating-fig-1b" class="level2">
<h2 class="anchored" data-anchor-id="recreating-fig-1b">Recreating Fig 1b</h2>
<p>Looking at outlies by genomic position in a Mannhatten plot</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Manhattan plot</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Prepare the data</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>Outliers <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">"Neutral"</span>, <span class="fu">length</span>(<span class="fu">colnames</span>(AllFreq)))</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>Outliers[<span class="fu">colnames</span>(AllFreq)<span class="sc">%in%</span>outliers<span class="sc">$</span>Loci] <span class="ot">&lt;-</span> <span class="st">"All outliers"</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>Outliers[<span class="fu">colnames</span>(AllFreq)<span class="sc">%in%</span>outliers_rdadapt_env] <span class="ot">&lt;-</span> <span class="st">"Top outliers"</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>Outliers <span class="ot">&lt;-</span> <span class="fu">factor</span>(Outliers, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Neutral"</span>, <span class="st">"All outliers"</span>, <span class="st">"Top outliers"</span>))</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>TAB_manhatan <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">pos =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">colnames</span>(AllFreq)), </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>                           <span class="at">pvalues =</span> rdadapt_env<span class="sc">$</span>p.values, </span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>                           <span class="at">Outliers =</span> Outliers)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>TAB_manhatan <span class="ot">&lt;-</span> TAB_manhatan[<span class="fu">order</span>(TAB_manhatan<span class="sc">$</span>Outliers),]</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot </span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> TAB_manhatan) <span class="sc">+</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x=</span>pos, <span class="at">y=</span><span class="sc">-</span><span class="fu">log10</span>(pvalues), <span class="at">col =</span> Outliers), <span class="at">size=</span><span class="fl">1.4</span>) <span class="sc">+</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"gray90"</span>, <span class="st">"#F9A242FF"</span>, <span class="st">"#6B4596FF"</span>)) <span class="sc">+</span></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Loci"</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"-log10(p.values)"</span>) <span class="sc">+</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept=</span><span class="sc">-</span><span class="fu">log10</span>(thres_env), <span class="at">linetype=</span><span class="st">"dashed"</span>, <span class="at">color =</span> <span class="fu">gray</span>(.<span class="dv">80</span>), <span class="at">size=</span><span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span><span class="st">"Manhattan plot"</span>, <span class="at">nrow =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color=</span><span class="fu">guide_legend</span>(<span class="at">title=</span><span class="st">"Locus type"</span>)) <span class="sc">+</span></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">11</span>, <span class="at">base_family =</span> <span class="st">"Times"</span>) <span class="sc">+</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"right"</span>, <span class="at">legend.background =</span> <span class="fu">element_blank</span>(), <span class="at">panel.grid =</span> <span class="fu">element_blank</span>(), <span class="at">legend.box.background =</span> <span class="fu">element_blank</span>(), <span class="at">plot.background =</span> <span class="fu">element_blank</span>(), <span class="at">panel.background =</span> <span class="fu">element_blank</span>(), <span class="at">legend.text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(.<span class="dv">8</span>)), <span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">11</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>We identified 557 loci showing extreme association with the environment, reduced to a set of 171 unlinked outliers when retaining only the best hit for each genomic contig.</p>
</blockquote>
</section>
<section id="simple-rda-for-outlier-identification" class="level2">
<h2 class="anchored" data-anchor-id="simple-rda-for-outlier-identification">Simple RDA for outlier identification</h2>
<blockquote class="blockquote">
<p>In this example, we accounted for population structure, as is commonly recommended when conducting genome scans; however, there are cases where this approach can be overly conservative. To investigate this issue in the lodgepole pine data, we compared the candidates identified with the partial RDA to those identified using a simple RDA.</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Running a simple RDA model</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>RDA_env_unconstrained <span class="ot">&lt;-</span> <span class="fu">rda</span>(AllFreq <span class="sc">~</span> MAR <span class="sc">+</span> EMT <span class="sc">+</span> MWMT <span class="sc">+</span> CMD <span class="sc">+</span> Tave_wt <span class="sc">+</span> DD_18 <span class="sc">+</span> MAP <span class="sc">+</span> Eref <span class="sc">+</span> PAS,  Variables)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Running the rdadapt function</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>rdadapt_env_unconstrained <span class="ot">&lt;-</span> <span class="fu">rdadapt</span>(RDA_env_unconstrained, <span class="dv">2</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Setting the p-value threshold </span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>thres_env <span class="ot">&lt;-</span> <span class="fl">0.01</span><span class="sc">/</span><span class="fu">length</span>(rdadapt_env_unconstrained<span class="sc">$</span>p.values)</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="do">## Identifying the outliers for the simple RDA</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>outliers_unconstrained <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Loci =</span> <span class="fu">colnames</span>(AllFreq)[<span class="fu">which</span>(rdadapt_env_unconstrained<span class="sc">$</span>p.values<span class="sc">&lt;</span>thres_env)], <span class="at">p.value =</span> rdadapt_env_unconstrained<span class="sc">$</span>p.values[<span class="fu">which</span>(rdadapt_env_unconstrained<span class="sc">$</span>p.values<span class="sc">&lt;</span>thres_env)], <span class="at">contig =</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(<span class="fu">colnames</span>(AllFreq)[<span class="fu">which</span>(rdadapt_env_unconstrained<span class="sc">$</span>p.values<span class="sc">&lt;</span>thres_env)], <span class="at">split =</span> <span class="st">"_"</span>), <span class="cf">function</span>(x) x[<span class="dv">1</span>])))</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>outliers_unconstrained <span class="ot">&lt;-</span> outliers_unconstrained[<span class="fu">order</span>(outliers_unconstrained<span class="sc">$</span>contig, outliers_unconstrained<span class="sc">$</span>p.value),]</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>outliers_rdadapt_env_unconstrained <span class="ot">&lt;-</span> <span class="fu">as.character</span>(outliers_unconstrained<span class="sc">$</span>Loci[<span class="sc">!</span><span class="fu">duplicated</span>(outliers_unconstrained<span class="sc">$</span>contig)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="overlap-among-all-outliers-with-and-without-conditioning-on-population-structure" class="level3">
<h3 class="anchored" data-anchor-id="overlap-among-all-outliers-with-and-without-conditioning-on-population-structure">Overlap among all outliers with and without conditioning on population structure</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="do">## For all the outliers</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>list_outliers_RDA_all <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">RDA_constrained =</span> <span class="fu">as.character</span>(outliers<span class="sc">$</span>Loci), <span class="at">RDA_unconstrained =</span> <span class="fu">as.character</span>(outliers_unconstrained<span class="sc">$</span>Loci))</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">#plot Venn diagram</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggVennDiagram</span>(list_outliers_RDA_all, <span class="at">category.names =</span> <span class="fu">c</span>(<span class="st">"partial RDA"</span>, <span class="st">"simple RDA"</span>), <span class="at">lty=</span><span class="st">"solid"</span>, <span class="at">size=</span><span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient2</span>(<span class="at">low =</span> <span class="st">"white"</span>, <span class="at">high =</span> <span class="st">'gray40'</span>) <span class="sc">+</span> <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"grey"</span>, <span class="st">"grey"</span>, <span class="st">"grey"</span>, <span class="st">"grey"</span>)) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">fill =</span> <span class="st">"none"</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">16</span>, <span class="at">family =</span> <span class="st">"Times"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>(Numbers don’t quite match the publication - note that the authors did publish a short correction that maybe takes care of this)</p>
</section>
<section id="overlap-among-rda-top-outliers-with-and-without-conditioning-on-population-structure" class="level3">
<h3 class="anchored" data-anchor-id="overlap-among-rda-top-outliers-with-and-without-conditioning-on-population-structure">Overlap among RDA (“top”) outliers with and without conditioning on population structure</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Only for the top hit locus per contig</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>list_outliers_RDA_top <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">RDA_constrained =</span> outliers_rdadapt_env, <span class="at">RDA_unconstrained =</span> outliers_rdadapt_env_unconstrained)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggVennDiagram</span>(list_outliers_RDA_top, <span class="at">category.names =</span> <span class="fu">c</span>(<span class="st">"partial RDA"</span>, <span class="st">"simple RDA"</span>), <span class="at">lty=</span><span class="st">"solid"</span>, <span class="at">size=</span><span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient2</span>(<span class="at">low =</span> <span class="st">"white"</span>, <span class="at">high =</span> <span class="st">'gray40'</span>) <span class="sc">+</span> <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"grey"</span>, <span class="st">"grey"</span>, <span class="st">"grey"</span>, <span class="st">"grey"</span>)) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">fill =</span> <span class="st">"none"</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">16</span>, <span class="at">family =</span> <span class="st">"Times"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>For all of the following RDA-based analyses we decided to retain only the 91 loci that were detected by both the partial RDA that accounted for population structure and by the simple RDA that did not account for population structure.</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>common_outliers_RDA_top <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(intersect, list_outliers_RDA_top)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p><strong>Notes on interpretation and best practices:</strong> We emphasize that there is no correct answer to the common dilemma of confounded variation in GEA, differentiation-based, and genome-wide association analyses. Instead, researchers must balance their tolerance for false negative and false positive rates, and use an analytical approach that best matches the objectives of the study as well as the potential ability to validate candidate loci. More generally, awareness of the tradeoffs associated with decisions such as population structure correction will allow users to appropriately interpret their results, both statistically and biologically. Decisions about if and how to correct for population structure, how to set outlier thresholds, and how to finalize a set of candidate adaptive markers will depend upon the objectives of the study and the data available (if any) for validation. The approach presented here is not meant to be applied without modification; modeling parameters (including what environmental predictors are used and how they are selected, see variable selection above) and other decisions should be determined based on the ecology of the organism, characteristics of the data set, and goals of the study.</p>
</blockquote>
</section>
</section>
<section id="spatial-projection-to-make-map-figures" class="level2">
<h2 class="anchored" data-anchor-id="spatial-projection-to-make-map-figures">Spatial projection to make map figures</h2>
<p>I was really hoping that I could get C&amp;F’s code working to allow you to recreate their Fig 3. Unfortunately the combination of changes to spatial packages and data files that are not completely reproducible has made this task much more difficult than I anticipated.</p>
<p>If you are interested in undertaking these analyses in the future, you are welcome to reach out to me. I will be setting up similar analyses in my own research and thus should have working examples in a few months - I would be happy to share examples and code with participants in this course. - Cynthia</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>