<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Land, River and Seascape Genomics - Intro to RDA as a general toolkit</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Land, River and Seascape Genomics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./Teachers.html" rel="" target="">
 <span class="menu-text">Meet your teachers</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./Readings.html" rel="" target="">
 <span class="menu-text">Readings</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./genetic-dif.html">Tuesday</a></li><li class="breadcrumb-item"><a href="./RDA.html">Intro to RDA as a general toolkit</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Sunday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Basic-Population-Genetics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic pop gen</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./basicR.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./basic-bash.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic Bash</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Monday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./land-river-sea.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Land, river, &amp; seascape genomics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./maps-spatial-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Maps &amp; spatial data</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Tuesday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genetic-dif.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Genetic differentiation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./RDA.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Intro to RDA as a general toolkit</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Wednesday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sims_demo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Sims &amp; demograph analyses</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Thursday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resistance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Resistance surfaces</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./GDM.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Future projections with GDM &amp; GFs</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./biophysical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Biophysical models</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">Friday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./GEA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">More GEA: genetic offsets &amp; maladaptation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./g_architectures.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Genomic architectures &amp; landscape genomics</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#geas-with-rda" id="toc-geas-with-rda" class="nav-link active" data-scroll-target="#geas-with-rda">GEAs with RDA</a>
  <ul class="collapse">
  <li><a href="#geas-genotype-by-environment-associations" id="toc-geas-genotype-by-environment-associations" class="nav-link" data-scroll-target="#geas-genotype-by-environment-associations">GEAs: Genotype-by-environment associations</a></li>
  </ul></li>
  <li><a href="#why-rda-and-other-multivariate-methods" id="toc-why-rda-and-other-multivariate-methods" class="nav-link" data-scroll-target="#why-rda-and-other-multivariate-methods">Why RDA (and other multivariate methods)?</a></li>
  <li><a href="#review-of-unconstrained-ordination-pca-pcoa" id="toc-review-of-unconstrained-ordination-pca-pcoa" class="nav-link" data-scroll-target="#review-of-unconstrained-ordination-pca-pcoa">Review of unconstrained ordination (PCA &amp; PCoA)</a>
  <ul class="collapse">
  <li><a href="#pca---principal-components-analysis" id="toc-pca---principal-components-analysis" class="nav-link" data-scroll-target="#pca---principal-components-analysis">PCA - principal components analysis</a></li>
  <li><a href="#pcoa-or-pco-principal-coordinates-analysis" id="toc-pcoa-or-pco-principal-coordinates-analysis" class="nav-link" data-scroll-target="#pcoa-or-pco-principal-coordinates-analysis">PCoA or PCO: Principal coordinates analysis</a></li>
  </ul></li>
  <li><a href="#simple-redundancy-analysis-constrained-ordination" id="toc-simple-redundancy-analysis-constrained-ordination" class="nav-link" data-scroll-target="#simple-redundancy-analysis-constrained-ordination">Simple redundancy analysis (constrained ordination)</a>
  <ul class="collapse">
  <li><a href="#example-from-capblancq-forester-2021---lodgepole-pine" id="toc-example-from-capblancq-forester-2021---lodgepole-pine" class="nav-link" data-scroll-target="#example-from-capblancq-forester-2021---lodgepole-pine">Example from Capblancq &amp; Forester 2021 - lodgepole pine</a>
  <ul class="collapse">
  <li><a href="#goals" id="toc-goals" class="nav-link" data-scroll-target="#goals">Goals</a></li>
  </ul></li>
  <li><a href="#data-preparation-for-rda" id="toc-data-preparation-for-rda" class="nav-link" data-scroll-target="#data-preparation-for-rda">Data preparation for RDA</a></li>
  <li><a href="#variable-selection-with-forward-model-building" id="toc-variable-selection-with-forward-model-building" class="nav-link" data-scroll-target="#variable-selection-with-forward-model-building">Variable selection with forward model building</a></li>
  </ul></li>
  <li><a href="#partial-rda" id="toc-partial-rda" class="nav-link" data-scroll-target="#partial-rda">partial RDA</a></li>
  <li><a href="#complex-models-and-variance-partitioning" id="toc-complex-models-and-variance-partitioning" class="nav-link" data-scroll-target="#complex-models-and-variance-partitioning">Complex models and variance partitioning</a></li>
  <li><a href="#extra-topic---dealing-with-spatial-correlation-structures" id="toc-extra-topic---dealing-with-spatial-correlation-structures" class="nav-link" data-scroll-target="#extra-topic---dealing-with-spatial-correlation-structures">Extra topic - dealing with spatial correlation structures</a>
  <ul class="collapse">
  <li><a href="#further-resources" id="toc-further-resources" class="nav-link" data-scroll-target="#further-resources">Further resources</a></li>
  <li><a href="#notes-developed-from" id="toc-notes-developed-from" class="nav-link" data-scroll-target="#notes-developed-from">Notes developed from:</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Intro to RDA as a general toolkit</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="geas-with-rda" class="level1">
<h1>GEAs with RDA</h1>
<section id="geas-genotype-by-environment-associations" class="level2">
<h2 class="anchored" data-anchor-id="geas-genotype-by-environment-associations">GEAs: Genotype-by-environment associations</h2>
<p>( = environmental associations)</p>
<ol type="1">
<li>Genome-wide: related to demographic history, ecological speciation/diversification, isolation by environment</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/IsolationByEnvironment1-Wang_and_Bradburd_2014.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Processes leading to IBE - Wang &amp; Bradburd 2014</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/IsolationByEnvironment2-Wang_and_Bradburd_2014.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">IBE signals - Wang &amp; Bradburd 2014</figcaption>
</figure>
</div>
<p><em>We will focus on this aspect today</em></p>
<ol start="2" type="1">
<li>Finding loci contribute to heritable genetic variation of selected traits</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/environment_genotype_phenotype_Rellstab2015.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Rellstab et al 2015 (Image modified from Sork et al 2013)</figcaption>
</figure>
</div>
<p>There is a huge literature on the topic of finding candidate loci for environmental selection that we cannot cover in one week. Population genetic tests of selection such as <em>outlier tests</em> or <em>genomic scans</em> are frequently used. There are also tests of selection that specifically look for associations of individual loci to environmental attributes - this will be more of our focus given that we are studying landscape genomics. We will come back to this topic on Friday. (Rellstab et al.&nbsp;2015 is a good review of the various methods if you are looking for further reading on this topic.)</p>
</section>
</section>
<section id="why-rda-and-other-multivariate-methods" class="level1">
<h1>Why RDA (and other multivariate methods)?</h1>
<p>RDA allows you to have multiple response variables (loci) and multiple predictors (spatial or environmental attributes)</p>
<table class="table">
<colgroup>
<col style="width: 28%">
<col style="width: 26%">
<col style="width: 44%">
</colgroup>
<thead>
<tr class="header">
<th>Predictive variables</th>
<th>Response variables</th>
<th>Appropriate family of methods *</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>One</td>
<td>One</td>
<td>Univariate/Simple regression</td>
</tr>
<tr class="even">
<td>Many</td>
<td>One</td>
<td>Multiple regression</td>
</tr>
<tr class="odd">
<td>Many</td>
<td>Many</td>
<td>Multivariate regression</td>
</tr>
</tbody>
</table>
<p>* Note that your genetic data not conform to expectations of parametric methods, including independence of data points</p>
<p>RDA was developed for community ecology data and has been adopted for genetic analyses. In an ecological analysis, there will be various spatial and environmental predictive variables (<em>n</em> sites by <em>m</em> predictive varaibles) and biological response variables (typically, species abundance organised in <em>n</em> sites by <em>p</em> species). Much of the literature about using ordination and multivariate analyses for environmental analyses will be described as ecological communities).</p>
<p>A key element of multivariate methods (including the RDA family) is <strong><em>ordination</em></strong> (which falls within the broader descriptor of eigenanalysis).</p>
<p>Ordination allows</p>
<ul>
<li><p>Compression and visualization of complex data</p></li>
<li><p>Flexibility to use distances to describe spatial and biological distances (between populations or individuals)</p></li>
<li><p>Flexibility to use allele frequencies (populations) or genotypes (individuals)</p></li>
</ul>
<p>Pitfalls on using ordination with genetic data</p>
<ul>
<li><p>Missing observations need to be excluded or imputed</p></li>
<li><p>Uneven sampling can skew results</p></li>
<li><p>Ordination is pattern description: many processes can yield the same patterns</p></li>
</ul>
<p>We focus on RDA as a multivariate method because</p>
<ul>
<li>it is flexible structure is very useful for landscape genomic data</li>
<li>it can describe genome-wide patterns</li>
<li>RDA performs well in genotype-environment-association analyses with low false positives and high true positives</li>
<li>The regression on one matrix on another.</li>
<li>Uses a combination of linear regression and PCA.</li>
</ul>
<p>RDA can address the following questions (citing Capblancq et al.&nbsp;2021):</p>
<p>1. What environmental/spatial processes drive patterns of genetic variation?</p>
<p>2. What is the genetic basis of local adaptation to the environment?</p>
<p>3. How is adaptive genetic variation distributed across landscapes?</p>
<p>4. What are the impacts of climate and/or landscape change on the distribution of adaptive genetic variation?</p>
<p><em>RDA takes linear combinations of the explanatory variables (X) and uses them to maximise the variance explained in linear combinations of Y (where X and Y are matrices)</em></p>
<ul>
<li>Keep this linearity in mind: it might not be appropriate for your data</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/RDA_overview_Capblancq_etal2021.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Conceptual overview of RDA - From Capblancq et al.&nbsp;2021</figcaption>
</figure>
</div>
</section>
<section id="review-of-unconstrained-ordination-pca-pcoa" class="level1">
<h1>Review of unconstrained ordination (PCA &amp; PCoA)</h1>
<section id="pca---principal-components-analysis" class="level2">
<h2 class="anchored" data-anchor-id="pca---principal-components-analysis">PCA - principal components analysis</h2>
<ul>
<li><p>No predictive variables: unconstrained.</p></li>
<li><p>Fits a line through the data along an axis that maximizes the variance described by the first PC axis (PC1).</p></li>
<li><p>Fits the next line to maximize the variance described for the second PC axis (PC2) where PC1 and PC2 are orthogonal (their variances are not correlated).</p></li>
<li><p>And so on.</p></li>
<li><p>Note that this is a linear procedure: lines are being fitted.</p></li>
<li><p>Eigenvalues (<span class="math inline">\(\lambda\)</span>) describe the amount of variance explained by each eigenvector (= line = PC axis)</p></li>
</ul>
</section>
<section id="pcoa-or-pco-principal-coordinates-analysis" class="level2">
<h2 class="anchored" data-anchor-id="pcoa-or-pco-principal-coordinates-analysis">PCoA or PCO: Principal coordinates analysis</h2>
<ul>
<li><p>same as <em>metric multidimensional scaling</em></p></li>
<li><p>takes pairwise distances (<span class="math inline">\(d_{ij}\)</span>, sometimes called dissimilarities) between points and uses ordination to find axes that maximally explain dissimilarities</p></li>
<li><p>PCoA on Euclidean distances is the same as PCA</p></li>
<li><p>Sometimes you get negative eigenvalues and have to use a correction</p></li>
<li><p>(non metric multidimensional scaling is similar but does not require linear data)</p></li>
</ul>
</section>
</section>
<section id="simple-redundancy-analysis-constrained-ordination" class="level1">
<h1>Simple redundancy analysis (constrained ordination)</h1>
<section id="example-from-capblancq-forester-2021---lodgepole-pine" class="level2">
<h2 class="anchored" data-anchor-id="example-from-capblancq-forester-2021---lodgepole-pine">Example from Capblancq &amp; Forester 2021 - lodgepole pine</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/istockphoto-149019996-612x612.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Credit: Kevin Cass</figcaption>
</figure>
</div>
<ul>
<li>use ~3000 “intergenic” SNPs as neutral loci (genotyped from 50K SNP chip)</li>
<li>geography, climate and neutral structure are all confounded</li>
</ul>
<section id="goals" class="level3">
<h3 class="anchored" data-anchor-id="goals">Goals</h3>
<ul>
<li>to undertake a GEA and ignore effect of geography+structure -&gt; possible false positives</li>
<li>to undertake a GEA and remove geography+structure -&gt; possible false negatives</li>
<li>There is no clear solution to the problems of false positives and false negatives</li>
</ul>
</section>
</section>
<section id="data-preparation-for-rda" class="level2">
<h2 class="anchored" data-anchor-id="data-preparation-for-rda">Data preparation for RDA</h2>
<ul>
<li>get environmental variables and scale them to a mean of 0, SD = 1</li>
<li>categorical predictors can be used as dummy variables (0/1)</li>
<li>look for colinearity among variables - might remove variables with variance inflation factor greater than ?</li>
</ul>
</section>
<section id="variable-selection-with-forward-model-building" class="level2">
<h2 class="anchored" data-anchor-id="variable-selection-with-forward-model-building">Variable selection with forward model building</h2>
<p>Build up a predictive model and assists with variable reduction</p>
<ol type="1">
<li><p>Test significance of <em>global model</em> (= all variables)</p></li>
<li><p>Start with “empty” model (= intercept only) and sequentially add variables</p></li>
<li><p>Two stopping criteria to avoid overfitting - end with either criterion</p>
<ul>
<li><p>permutation based significance test</p></li>
<li><p>adjusted <span class="math inline">\(R^2\)</span> from global model</p></li>
</ul></li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/simpleRDA_Capblancq_and_Forester_2022.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Simple RDA output - Capblancq &amp; Forester 2021</figcaption>
</figure>
</div>
</section>
</section>
<section id="partial-rda" class="level1">
<h1>partial RDA</h1>
<p>Allows independent estimation of sets of variables together with confounded effects caused by collinearity.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/partialRDA__Capblancq_etal2021.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Reporting on partial RDA-From Capblancq et al.&nbsp;2021</figcaption>
</figure>
</div>
</section>
<section id="complex-models-and-variance-partitioning" class="level1">
<h1>Complex models and variance partitioning</h1>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/summarising_complex_models_Zbinden_etal_2023.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Models upon models - Zbinden et al.&nbsp;2023</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/partionedvariance1_Zbinden_etal_2023.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Variance partitioning - averages across species - Zbinden et al 2023</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/partionedvariance2_Zbinden_etal_2023.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Variance partitioning - species details - Zbinden et al 2023</figcaption>
</figure>
</div>
</section>
<section id="extra-topic---dealing-with-spatial-correlation-structures" class="level1">
<h1>Extra topic - dealing with spatial correlation structures</h1>
<p>Variety of approaches (to be explored in the computer activities that follow)</p>
<ul>
<li><p>Include spatial position as a covariate</p></li>
<li><p>Use genetic distances and linearize with PCoA</p></li>
<li><p>Use Moran Eigenvector Mapping</p></li>
</ul>
<section id="further-resources" class="level2">
<h2 class="anchored" data-anchor-id="further-resources">Further resources</h2>
<ul>
<li><p>The absolutely go-to package in R for all ecological eigenanalysis is <code>vegan</code>. It is very well documented and has excellent <a href="https://cran.r-project.org/web/packages/vegan/index.html">vignettes</a> that are well-worth working through.</p></li>
<li><p>Really nice slide deck that demonstrates some vegan functions with excellent illustrations <a href="https://fromthebottomoftheheap.net/slides/intro-vegan-webinar-2020/intro-to-vegan.html#1">Intro to vegan</a></p></li>
<li><p>Another great slide deck that runs through RDA and associated vegan functions <a href="https://pmassicotte.github.io/stats-denmark-2019/07_rda.html#/">Redundancy Analysis</a></p></li>
</ul>
</section>
<section id="notes-developed-from" class="level2">
<h2 class="anchored" data-anchor-id="notes-developed-from">Notes developed from:</h2>
<blockquote class="blockquote">
<ul>
<li>Capblancq, T., &amp; Forester, B. R. (2021). Redundancy analysis: A Swiss Army Knife for landscape genomics. Methods in Ecology and Evolution. doi:10.1111/2041-210x.13722</li>
</ul>
</blockquote>
<blockquote class="blockquote">
<ul>
<li>Rellstab, C., Gugerli, F., Eckert, A. J., Hancock, A. M., &amp; Holderegger, R. (2015). A practical guide to environmental association analysis in landscape genomics. Molecular Ecology, 24 (17), 4348-4370.</li>
</ul>
</blockquote>
<blockquote class="blockquote">
<ul>
<li>Wang, I. J., &amp; Bradburd, G. S. (2014). Isolation by environment. Molecular Ecology, 23(23), 5649-5662. doi:papers3://publication/doi/10.1111/mec.12938</li>
</ul>
</blockquote>
<blockquote class="blockquote">
<ul>
<li>Zbinden, Z. D., Douglas, M. R., Chafin, T. K., &amp; Douglas, M. E. (2022). Riverscape community genomics: A comparative analytical approach to identify common drivers of spatial structure. Molecular Ecology, 32, XXX-XXX. doi:10.1111/mec.16806</li>
</ul>
</blockquote>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>