<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Land, River and Seascape Genomics - Readings</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Land, River and Seascape Genomics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./Teachers.html" rel="" target="">
 <span class="menu-text">Meet your teachers</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./Readings.html" rel="" target="" aria-current="page">
 <span class="menu-text">Readings</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Readings</li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Sunday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Basic-Population-Genetics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic pop gen</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./basicR.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./basic-bash.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic Bash</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Monday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./land-river-sea.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Land, river, sea</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./maps-spatial-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Maps &amp; spatial data</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Tuesday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genetic-dif.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Genetic differentiation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./RDA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Intro to GEA &amp; RDA as a general toolkit</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Wednesday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genetic-dif.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Genetic differentiation</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Thursday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resistance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Resistance surfaces</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./GDM.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Future projectsions with GDM &amp; GFs</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./biophysical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Biophysical models</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">Friday</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./GEA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">More GEA: genetic offsets &amp; maladaptation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./g_architectures.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Genomic architectures &amp; landscape genomics</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#readings-by-day-and-topic" id="toc-readings-by-day-and-topic" class="nav-link active" data-scroll-target="#readings-by-day-and-topic">Readings by day and topic</a>
  <ul class="collapse">
  <li><a href="#sunday" id="toc-sunday" class="nav-link" data-scroll-target="#sunday">Sunday</a></li>
  <li><a href="#monday" id="toc-monday" class="nav-link" data-scroll-target="#monday">Monday</a></li>
  <li><a href="#tuesday" id="toc-tuesday" class="nav-link" data-scroll-target="#tuesday">Tuesday</a></li>
  <li><a href="#wednesday" id="toc-wednesday" class="nav-link" data-scroll-target="#wednesday">Wednesday</a></li>
  <li><a href="#thursday" id="toc-thursday" class="nav-link" data-scroll-target="#thursday">Thursday</a></li>
  <li><a href="#friday" id="toc-friday" class="nav-link" data-scroll-target="#friday">Friday</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Readings</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="readings-by-day-and-topic" class="level2">
<h2 class="anchored" data-anchor-id="readings-by-day-and-topic">Readings by day and topic</h2>
<p>For each topic, there will be a required reading (or pre-class activity) and several optional readings.</p>
<p>We expect you to have read the required reading before the day. There will not be time during the week of the course to read new materials and therefore we <em>highly suggest</em> that you do these readings in advance of the course starting.</p>
<section id="sunday" class="level3">
<h3 class="anchored" data-anchor-id="sunday">Sunday</h3>
<section id="basic-population-genetics" class="level4">
<h4 class="anchored" data-anchor-id="basic-population-genetics">Basic Population Genetics</h4>
<p><em>No background reading is required for this module.</em></p>
</section>
<section id="basic-r" class="level4">
<h4 class="anchored" data-anchor-id="basic-r">Basic R</h4>
<p>If you are new to R, please attempt the following tutorials on your own. If you are experienced in R, skim through the tutorials and make sure that all the topics are familiar to you.</p>
<ul>
<li><p><a href="https://evolutionarygenetics.github.io/Introduction.html">Ravinet’s introduction to R</a></p></li>
<li><p><a href="https://evolutionarygenetics.github.io/Chapter2.html">Going further with R</a></p></li>
</ul>
<p>Through this course, most of the examples you are likely to encounter will predominately use base R. However, learning Tidyverse syntax will make your life easier in the long run, especially if you are going to be manipulating spatial data.</p>
</section>
<section id="basic-bash" class="level4">
<h4 class="anchored" data-anchor-id="basic-bash">Basic Bash</h4>
<p>As with R, you need to have some basic comfort and familiarity with bash. Either ensure you can complete the tutoirals on your own or attend this special Sunday session.</p>
<p>See activities and links on <a href="./basic-bash.html">bash tutorial webpage</a>.</p>
</section>
</section>
<section id="monday" class="level3">
<h3 class="anchored" data-anchor-id="monday">Monday</h3>
<section id="land-river-sea" class="level4">
<h4 class="anchored" data-anchor-id="land-river-sea">Land, river, sea</h4>
<section id="required" class="level5">
<h5 class="anchored" data-anchor-id="required">Required</h5>
<ul>
<li>Storfer, A., Patton, A., &amp; Fraik, A. K. (2018). Navigating the interface between landscape genetics and landscape genomics. <em>Front Genet, 9</em>, 68. doi:10.3389/fgene.2018.00068</li>
</ul>
</section>
<section id="optional" class="level5">
<h5 class="anchored" data-anchor-id="optional">Optional</h5>
<ul>
<li><p>Manel, S., Schwartz, M. K., Luikart, G., &amp; Taberlet, P. (2003). Landscape genetics: combining landscape ecology and population genetics. Trends in Ecology and Evolution, 18(4), 189-197. doi:papers3://publication/doi/10.1016/S0169-5347(03)00008-9 <em>This is the original paper defining landscape genetics as a field.</em></p></li>
<li><p>Find a good terrestrial overview</p></li>
<li><p>Riginos, C., &amp; Liggins, L. (2013). Seascape genetics: populations, individuals, and genes marooned and adrift. <em>Geography Compass, 7</em>(3), 197-216. doi:papers3://publication/doi/10.1111/gec3.12032 <em>Discusses many ways that marine landscape genetics is distinct from terrestrial</em></p></li>
<li><p>Selkoe, K. A., Scribner, K. T., &amp; Galindo, H. M. (2016). Waterscape genetics – applications of landscape genetics to rivers, lakes, and seas. In N. Balkenhol, S. A. Cushman, A. Storfer, &amp; L. P. Waits (Eds.), Landscape Genetics: Concepts, Methods, Applications (pp.&nbsp;1-27): John Wiley &amp; Sons, Ltd.&nbsp;<em>Includes overviews on some of the contrasting features of marine and freshwater environments</em></p></li>
<li><p>Liggins, L., Treml, E. A., &amp; Riginos, C. (2019). Seascape genomics: contextualizing adaptive and neutral genomic variation in the ocean environment. <em>Overviews on issues particular to marine studies</em></p></li>
<li><p>Blanchet, S., Prunier, J. G., Paz-Vinas, I., Saint-Pe, K., Rey, O., Raffard, A., Mathieu-Begne, E., Loot, G., Fourtune, L., &amp; Dubut, V. (2020). A river runs through it: The causes, consequences, and management of intraspecific diversity in river networks. Evolutionary Applications, 13(6), 1195-1213. doi:10.1111/eva.12941 <em>Features and emerging conclusions from freshwater systems</em></p></li>
</ul>
</section>
</section>
<section id="maps-spatial-data" class="level4">
<h4 class="anchored" data-anchor-id="maps-spatial-data">Maps &amp; spatial data</h4>
<p><em>No readings are required.</em></p>
<ul>
<li><p>The best website explaining how to use spatial data in R is [<em>Geocomputation with R</em>, a book on geographic data analysis, visualization and modeling](https://r.geocompx.org)</p></li>
<li><p>Leempoel, K., Duruz, S., Rochat, E., Widmer, I., Orozco-terWengel, P., &amp; Joost, S. (2017). Simple rules for an efficient use of geographic information systems in molecular ecology. _Frontiers in Ecology and Evolution, 5_. doi:10.3389/fevo.2017.00033 *Reviews various approaches and tools, good list of terrestrial environmental data sources included in appendix*</p></li>
</ul>
</section>
</section>
<section id="tuesday" class="level3">
<h3 class="anchored" data-anchor-id="tuesday">Tuesday</h3>
<section id="genetic-differentiation-and-genetic-structuring" class="level4">
<h4 class="anchored" data-anchor-id="genetic-differentiation-and-genetic-structuring">Genetic differentiation and genetic structuring</h4>
<section id="required-1" class="level5">
<h5 class="anchored" data-anchor-id="required-1">Required</h5>
<ul>
<li>Waples, R. S., &amp; Gaggiotti, O. E. (2006). What is a population? An empirical evaluation of some genetic methods for identifying the number of gene pools and their degree of connectivity. <em>Molecular Ecology, 15</em>(6), 1419-1439. doi:papers3://publication/doi/10.1111/j.1365-294X.2006.02890.x <em>This study gives a nice example of what a population is.</em></li>
<li>Leslie, S., Winney, B., Hellenthal, G., Davison, D., Boumertit, A., Day, T., Hutnik, K., Royrvik, E. C., Cunliffe, B., Wellcome Trust Case Control, C., International Multiple Sclerosis Genetics, C., Lawson, D. J., Falush, D., Freeman, C., Pirinen, M., Myers, S., Robinson, M., Donnelly, P., &amp; Bodmer, W. (2015). The fine-scale genetic structure of the British population. <em>Nature, 519</em>(7543), 309-314. doi:10.1038/nature14230 <em>This study is a nice example of how clustering can be used to study population structure.</em></li>
</ul>
</section>
<section id="optional-1" class="level5">
<h5 class="anchored" data-anchor-id="optional-1">Optional</h5>
<ul>
<li>Linck, E., &amp; Battey, C. J. (2019). Minor allele frequency thresholds strongly affect population structure inference with genomic data sets. <em>Molecular Ecology Resources, 19</em>(3), 639-647. doi:10.1111/1755-0998.12995</li>
</ul>
</section>
</section>
<section id="rda-as-a-general-toolkit" class="level4">
<h4 class="anchored" data-anchor-id="rda-as-a-general-toolkit">RDA as a general toolkit</h4>
<section id="required-2" class="level5">
<h5 class="anchored" data-anchor-id="required-2">Required</h5>
<ul>
<li>Capblancq, T., &amp; Forester, B. R. (2021). Redundancy analysis: A Swiss Army Knife for landscape genomics. Methods in Ecology and Evolution. doi:10.1111/2041-210x.13722 <em>Clear explanations and worked examples that we will follow in class</em></li>
</ul>
</section>
<section id="optional-2" class="level5">
<h5 class="anchored" data-anchor-id="optional-2">Optional</h5>
<ul>
<li><p>Rellstab, C., Gugerli, F., Eckert, A. J., Hancock, A. M., &amp; Holderegger, R. (2015). A practical guide to environmental association analysis in landscape genomics. Molecular Ecology, 24(17), 4348-4370. doi:papers3://publication/doi/10.1111/mec.13322 <em>Highly recommend reading as a general overview to GEA/EEA analyses and concepts</em></p></li>
<li><p>Forester, B. R., Jones, M. R., Joost, S., Landguth, E. L., &amp; Lasky, J. R. (2016). Detecting spatial genetic signatures of local adaptation in heterogeneous landscapes. Molecular Ecology, 25(1), 104-120. doi:papers3://publication/doi/10.1111/mec.13476 <em>Evaluates various methods for finding genotype-by-environment associations and finds that RDA methods often best</em></p></li>
<li><p>Lasky, J. R., Josephs, E. B., &amp; Morris, G. P. (2023). Genotype-environment associations to reveal the molecular basis of environmental adaptation. <em>Plant Cell, 35</em>, 125-138. doi:10.1093/plcell/koac267 <em>Focusing on plants, reviews the strengths and shortcomings of GEA approaches</em></p></li>
</ul>
</section>
</section>
</section>
<section id="wednesday" class="level3">
<h3 class="anchored" data-anchor-id="wednesday">Wednesday</h3>
<section id="simulations-and-demographic-analyses" class="level4">
<h4 class="anchored" data-anchor-id="simulations-and-demographic-analyses">Simulations and demographic analyses</h4>
<section id="required-3" class="level5">
<h5 class="anchored" data-anchor-id="required-3">Required</h5>
<ul>
<li><p>Sousa, V., Hey, J. Understanding the origin of species with genome-scale data: modelling gene flow.<em>Nat Rev Genet</em>&nbsp;<strong>14</strong>, 404–414 (2013). https://doi.org/10.1038/nrg3446</p></li>
<li><p>Kelleher, J., Wong, Y., Wohns, A.W.&nbsp;<em>et al.</em>&nbsp;Inferring whole-genome histories in large population datasets.&nbsp;<em>Nat Genet</em>&nbsp;<strong>51</strong>, 1330–1338 (2019). https://doi.org/10.1038/s41588-019-0483-y</p></li>
</ul>
</section>
<section id="optional-3" class="level5">
<h5 class="anchored" data-anchor-id="optional-3">Optional</h5>
<ul>
<li>Meier, J.I., Sousa, V.C., Marques, D.A., Selz, O.M., Wagner, C.E., Excoffier, L. and Seehausen, O. (2017), Demographic modelling with whole-genome data reveals parallel origin of similar Pundamilia cichlid species after hybridization. Mol Ecol, 26: 123-141. https://doi.org/10.1111/mec.13838</li>
</ul>
</section>
</section>
</section>
<section id="thursday" class="level3">
<h3 class="anchored" data-anchor-id="thursday">Thursday</h3>
<section id="resistance-surfaces" class="level4">
<h4 class="anchored" data-anchor-id="resistance-surfaces">Resistance surfaces</h4>
<section id="required-4" class="level5">
<h5 class="anchored" data-anchor-id="required-4">Required</h5>
<ul>
<li><p>Beninde, J., Wittische, J., &amp; Frantz, A. C. (2023). Quantifying uncertainty in inferences of landscape genetic resistance due to choice of individual-based genetic distance metric. <em>Molecular Ecology Resources</em>. doi:10.1111/1755-0998.13831</p></li>
<li><p>Peterman, W. E. (2018). ResistanceGA: An R package for the optimization of resistance surfaces using genetic algorithms. <em>Methods in Ecology and Evolution, 9</em>(6), 1638-1647. doi:papers3://publication/doi/10.1111/2041-210X.12984</p></li>
</ul>
</section>
</section>
<section id="future-projections-with-gdm-gfs" class="level4">
<h4 class="anchored" data-anchor-id="future-projections-with-gdm-gfs">Future projections with GDM &amp; GFs</h4>
<section id="required-5" class="level5">
<h5 class="anchored" data-anchor-id="required-5">Required</h5>
<ul>
<li>Fitzpatrick, M. C., &amp; Keller, S. R. (2015). Ecological genomics meets community-level modelling of biodiversity: mapping the genomic landscape of current and future environmental adaptation. Ecol Lett, 18(1), 1-16. doi:papers3://publication/doi/10.1111/ele.12376</li>
</ul>
</section>
<section id="optional-4" class="level5">
<h5 class="anchored" data-anchor-id="optional-4">Optional</h5>
<ul>
<li><p>Mokany, K., Ware, C., Woolley, S. N. C., Ferrier, S., Fitzpatrick, Matthew&nbsp;C., &amp; Bahn, V. (2022). A working guide to harnessing generalized dissimilarity modelling for biodiversity analysis and conservation assessment. <em>Global Ecology and Biogeography, 31</em>(4), 802-821. doi:10.1111/geb.13459. <em>A worked example for GDM</em></p></li>
<li><p>Ferrier, S., Manion, G., Elith, J., &amp; Richardson, K. (2007). Using generalized dissimilarity modelling to analyse and predict patterns of beta diversity in regional biodiversity assessment. <em>Diversity and Distributions, 13</em>(3), 252-264. doi:papers3://publication/doi/10.1111/j.1472-4642.2007.00341.x <em>Original method for GDM</em></p></li>
<li><p>Ellis, N., Smith, S. J., &amp; Pitcher, C. R. (2012). Gradient forests: calculating importance gradients on physical predictors. <em>Ecology, 93</em>(1), 156-168. doi:papers3://publication/uuid/06CCD6BA-6A4B-49EB-AE26-736E8A04B9C8 <em>Original method for GF</em></p></li>
</ul>
</section>
</section>
<section id="biophysical-models" class="level4">
<h4 class="anchored" data-anchor-id="biophysical-models">Biophysical models</h4>
<section id="required-6" class="level5">
<h5 class="anchored" data-anchor-id="required-6">Required</h5>
<ul>
<li>Jahnke, M., Jonsson, P. R., Moksnes, P. O., Loo, L. O., Nilsson Jacobi, M., &amp; Olsen, J. L. (2017). Seascape genetics and biophysical connectivity modelling support conservation of the seagrass <em>Zostera marina</em> in the Skagerrak-Kattegat region of the eastern North Sea. Evolutionary Applications, 1-46. doi:papers3://publication/doi/10.1111/eva.12589 <em>Comprehensive review of studies pairing seascape genetic data with biophysical models</em></li>
</ul>
</section>
<section id="optional-5" class="level5">
<h5 class="anchored" data-anchor-id="optional-5">Optional</h5>
<ul>
<li>Boulanger, E., Dalongeville, A., Andrello, M., Mouillot, D., &amp; Manel, S. (2020). Spatial graphs highlight how multi‐generational dispersal shapes landscape genetic patterns. Ecography. doi:10.1111/ecog.05024</li>
</ul>
</section>
</section>
</section>
<section id="friday" class="level3">
<h3 class="anchored" data-anchor-id="friday">Friday</h3>
<section id="more-gea-genetic-offsets-and-maladaptation" class="level4">
<h4 class="anchored" data-anchor-id="more-gea-genetic-offsets-and-maladaptation">More GEA: genetic offsets and maladaptation</h4>
<section id="required-7" class="level5">
<h5 class="anchored" data-anchor-id="required-7">Required</h5>
<ul>
<li>Capblancq, T., Fitzpatrick, M. C., Bay, R. A., Exposito-Alonso, M., &amp; Keller, S. R. (2020). Genomic prediction of (mal)adaptation across current and future climatic landscapes. <em>Annual Review of Ecology, Evolution, and Systematics, 51</em>(1), 245-269. doi:10.1146/annurev-ecolsys-020720-042553 <em>Review of maladaptation and genomic offset concepts</em></li>
</ul>
</section>
<section id="optional-6" class="level5">
<h5 class="anchored" data-anchor-id="optional-6">Optional</h5>
<ul>
<li><p>Gougherty, A. V., Keller, S. R., &amp; Fitzpatrick, M. C. (2021). Maladaptation, migration and extirpation fuel climate change risk in a forest tree species. <em>Nat Clim Change, 11</em>(2), 166-171. doi:10.1038/s41558-020-00968-6 <em>Interesting conceptual example that integrates genomic offsets with dispersal capacity</em></p></li>
<li><p>Fitzpatrick, M. C., V. E. Chhatre, R. Y. Soolanayakanahally and S. R. Keller, 2021 Experimental support for genomic prediction of climate maladaptation using the machine learning approach Gradient Forests. Mol Ecol Resour 21<strong>:</strong> 2749-2765. <em>One of the best examples of an empirical test of GEA predictions</em></p></li>
</ul>
</section>
</section>
<section id="landscape-genomics-and-genetic-architectures" class="level4">
<h4 class="anchored" data-anchor-id="landscape-genomics-and-genetic-architectures">Landscape genomics and genetic architectures</h4>
<p><em>No readings are required.</em></p>
<section id="optional-7" class="level5">
<h5 class="anchored" data-anchor-id="optional-7">Optional</h5>
<ul>
<li><p>Laruson, A. J., Fitzpatrick, M. C., Keller, S. R., Haller, B. C., &amp; Lotterhos, K. E. (2022). Seeing the forest for the trees: Assessing genetic offset predictions from gradient forest. <em>Evol Appl, 15</em>(3), 403-416. doi:10.1111/eva.13354</p></li>
<li><p>Lotterhos, K. E. (2019). The effect of neutral recombination variation on genome scans for selection. <em>G3 (Bethesda), 9</em>(6), 1851-1867. doi:10.1534/g3.119.400088</p></li>
<li><p>Bierne, N., Gagnaire, P.-A., &amp; David, P. (2013). The geography of introgression in a patchy environment and the thorn in the side of ecological speciation. <em>Current Zoology, 59</em>(1), 72-86. doi:papers3://publication/doi/10.1093/czoolo/59.1.72</p></li>
</ul>


</section>
</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>